{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Statistics tutorial**\n",
    "\n",
    "In this Jupyter tutorial we will show some examples of how to extract statistics from our data. These methods are implemented as part of the functionality of PyOskar API using the _transformers_ Spark model. These kind of functions depend directly from an _Oskar()_ object, and they will return a whole dataframe with the due transformation applied.\n",
    "> <span style=\"color:#ff6600\"> **PyOskar is integrated in PySpark default API. It is possible to manage data simultaneously with functions from both libraries in case we want to select specific fields, visualize the dataframe, filter the output or performing other operations besides the transformation.** </span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to import the PyOskar and PySpark modules. Second, we need to create an instance of the _Oskar()_ object, from which depends a big part of the functionality. Finally, we must use the _load()_ transformer pointing to where the parquet file is stored to convert our data into a DataFrame _df_, and we are ready to start playing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyoskar.core import Oskar\n",
    "from pyoskar.sql import *\n",
    "from pyoskar.analysis import *\n",
    "from pyspark.sql.functions import *\n",
    "\n",
    "oskar = Oskar(spark)\n",
    "df = oskar.load(\"./data/platinum_chr22.small.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can use PySpark _show()_ method to print the data from _df_. This is how our testing dataframe looks like. As you can see for this tutorial we have selected a small dataset from Illumina Platinum Genomes with 1,000 random variants from chromosome 22."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Print first 20 variants:\n",
      "+---------------+-----+----------+--------+--------+---------+---------+------+----+------+-----+----+--------------------+--------------------+\n",
      "|             id|names|chromosome|   start|     end|reference|alternate|strand|  sv|length| type|hgvs|             studies|          annotation|\n",
      "+---------------+-----+----------+--------+--------+---------+---------+------+----+------+-----+----+--------------------+--------------------+\n",
      "|22:16054454:C:T|   []|        22|16054454|16054454|        C|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16054454, 16...|\n",
      "|22:16065809:T:C|   []|        22|16065809|16065809|        T|        C|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16065809, 16...|\n",
      "|22:16077310:T:A|   []|        22|16077310|16077310|        T|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16077310, 16...|\n",
      "|22:16080499:A:G|   []|        22|16080499|16080499|        A|        G|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16080499, 16...|\n",
      "|22:16084621:T:C|   []|        22|16084621|16084621|        T|        C|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16084621, 16...|\n",
      "|22:16091610:G:T|   []|        22|16091610|16091610|        G|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16091610, 16...|\n",
      "|22:16096040:G:A|   []|        22|16096040|16096040|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16096040, 16...|\n",
      "|22:16099957:C:T|   []|        22|16099957|16099957|        C|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16099957, 16...|\n",
      "|22:16100462:A:G|   []|        22|16100462|16100462|        A|        G|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16100462, 16...|\n",
      "|22:16105660:G:A|   []|        22|16105660|16105660|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16105660, 16...|\n",
      "|22:16112391:G:A|   []|        22|16112391|16112391|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16112391, 16...|\n",
      "|22:16114913:A:T|   []|        22|16114913|16114913|        A|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16114913, 16...|\n",
      "|22:16127471:A:-|   []|        22|16127471|16127471|        A|         |     +|null|     1|INDEL|  []|[[hgvauser@platin...|[22, 16127471, 16...|\n",
      "|22:16134019:G:T|   []|        22|16134019|16134019|        G|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16134019, 16...|\n",
      "|22:16138943:C:G|   []|        22|16138943|16138943|        C|        G|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16138943, 16...|\n",
      "|22:16144239:T:C|   []|        22|16144239|16144239|        T|        C|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16144239, 16...|\n",
      "|22:16147398:G:A|   []|        22|16147398|16147398|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16147398, 16...|\n",
      "|22:16149692:G:T|   []|        22|16149692|16149692|        G|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16149692, 16...|\n",
      "|22:16195955:G:A|   []|        22|16195955|16195955|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16195955, 16...|\n",
      "|22:16196041:C:T|   []|        22|16196041|16196041|        C|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16196041, 16...|\n",
      "+---------------+-----+----------+--------+--------+---------+---------+------+----+------+-----+----+--------------------+--------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Print first 20 variants:\")\n",
    "df.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOTE: With _df.printSchema()_ command you can check the dataset hierarchy and all its fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of variants:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1000"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Total number of variants:\")\n",
    "df.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats per variant\n",
    "Here we present the genotypes of the variants from which the later stats will come out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "|id             |genotypes                                                                                                                |\n",
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "|22:16054454:C:T|[[[./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [./.], [./.], [./.]]]|\n",
      "|22:16065809:T:C|[[[0/1], [./.], [./.], [./.], [0/1], [./.], [0/1], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [./.], [./.], [0/1]]]|\n",
      "|22:16077310:T:A|[[[./.], [0/1], [0/1], [./.], [./.], [0/1], [./.], [0/1], [./.], [0/1], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16080499:A:G|[[[0/1], [./.], [0/1], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [./.], [./.], [./.]]]|\n",
      "|22:16084621:T:C|[[[./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16091610:G:T|[[[./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16096040:G:A|[[[./.], [./.], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [0/1]]]|\n",
      "|22:16099957:C:T|[[[0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [1/1], [0/1], [1/1], [1/1], [0/1], [0/1]]]|\n",
      "|22:16100462:A:G|[[[0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1]]]|\n",
      "|22:16105660:G:A|[[[0/1], [./.], [0/1], [./.], [0/1], [./.], [0/1], [0/1], [./.], [0/1], [./.], [0/1], [0/1], [./.], [0/1], [./.], [0/1]]]|\n",
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.select(df.id, df.studies.samplesData.alias(\"genotypes\")).show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case we are applying a transformation that adds new data inside the _stats_ field, which is empty before we do anything:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16065809:T:C|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16077310:T:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16080499:A:G|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16084621:T:C|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16091610:G:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16096040:G:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16099957:C:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16100462:A:G|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16105660:G:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\")\\\n",
    "    .show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we proceed to call the _stats()_ transformer through our _oskar_ instance. That means that our dataframe will go through some kind of operation. It is important to know how transformers work in order to correctly access the new information.\n",
    "<br>\n",
    "Usage:\n",
    "```python\n",
    "stats(df[DataFrame], studyId[str]=None, cohort[str]=None, samples[str]=None, missingAsReference[bool]=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|2          |1             |1             |32                |16                  |0.5          |0.5          |\n",
      "|22:16065809:T:C|12         |6             |6             |22                |11                  |0.5          |0.5          |\n",
      "|22:16077310:T:A|14         |7             |7             |20                |10                  |0.5          |0.5          |\n",
      "|22:16080499:A:G|8          |4             |4             |26                |13                  |0.5          |0.5          |\n",
      "|22:16084621:T:C|10         |5             |5             |24                |12                  |0.5          |0.5          |\n",
      "|22:16091610:G:T|6          |3             |3             |28                |14                  |0.5          |0.5          |\n",
      "|22:16096040:G:A|18         |9             |9             |16                |8                   |0.5          |0.5          |\n",
      "|22:16099957:C:T|34         |14            |20            |0                 |0                   |0.4117647    |0.5882353    |\n",
      "|22:16100462:A:G|32         |16            |16            |2                 |1                   |0.5          |0.5          |\n",
      "|22:16105660:G:A|20         |10            |10            |14                |7                   |0.5          |0.5          |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "statsDF = oskar.stats(df)\n",
    "statsDF.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\")\\\n",
    "    .show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since our data has some empty genotypes, we would like to treat those missing values as if they were the reference. For that issue we dispose of _missingAsReference_ parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|34         |33            |1             |0                 |0                   |0.9705882    |0.029411765  |\n",
      "|22:16065809:T:C|34         |28            |6             |0                 |0                   |0.8235294    |0.1764706    |\n",
      "|22:16077310:T:A|34         |27            |7             |0                 |0                   |0.7941176    |0.20588236   |\n",
      "|22:16080499:A:G|34         |30            |4             |0                 |0                   |0.88235295   |0.11764706   |\n",
      "|22:16084621:T:C|34         |29            |5             |0                 |0                   |0.85294116   |0.14705883   |\n",
      "|22:16091610:G:T|34         |31            |3             |0                 |0                   |0.9117647    |0.0882353    |\n",
      "|22:16096040:G:A|34         |25            |9             |0                 |0                   |0.7352941    |0.2647059    |\n",
      "|22:16099957:C:T|34         |14            |20            |0                 |0                   |0.4117647    |0.5882353    |\n",
      "|22:16100462:A:G|34         |18            |16            |0                 |0                   |0.5294118    |0.47058824   |\n",
      "|22:16105660:G:A|34         |24            |10            |0                 |0                   |0.7058824    |0.29411766   |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "missingAsReferenceStatsDF = oskar.stats(df, missingAsReference = True)\n",
    "missingAsReferenceStatsDF.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\")\\\n",
    "    .show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Other statistics that our method provides are the next ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "|id             |genotypeCount        |genotypeFreq                                     |mafAllele|maf        |mgfGenotype|mgf|\n",
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "|22:16054454:C:T|[0/0 -> 16, 0/1 -> 1]|[0/0 -> 0.9411765, 0/1 -> 0.05882353, 1/1 -> 0.0]|T        |0.029411765|1/1        |0.0|\n",
      "|22:16065809:T:C|[0/0 -> 11, 0/1 -> 6]|[0/0 -> 0.64705884, 0/1 -> 0.3529412, 1/1 -> 0.0]|C        |0.1764706  |1/1        |0.0|\n",
      "|22:16077310:T:A|[0/0 -> 10, 0/1 -> 7]|[0/0 -> 0.5882353, 0/1 -> 0.4117647, 1/1 -> 0.0] |A        |0.20588236 |1/1        |0.0|\n",
      "|22:16080499:A:G|[0/0 -> 13, 0/1 -> 4]|[0/0 -> 0.7647059, 0/1 -> 0.23529412, 1/1 -> 0.0]|G        |0.11764706 |1/1        |0.0|\n",
      "|22:16084621:T:C|[0/0 -> 12, 0/1 -> 5]|[0/0 -> 0.7058824, 0/1 -> 0.29411766, 1/1 -> 0.0]|C        |0.14705883 |1/1        |0.0|\n",
      "|22:16091610:G:T|[0/0 -> 14, 0/1 -> 3]|[0/0 -> 0.8235294, 0/1 -> 0.1764706, 1/1 -> 0.0] |T        |0.0882353  |1/1        |0.0|\n",
      "|22:16096040:G:A|[0/0 -> 8, 0/1 -> 9] |[0/0 -> 0.47058824, 0/1 -> 0.5294118, 1/1 -> 0.0]|A        |0.2647059  |1/1        |0.0|\n",
      "|22:16099957:C:T|[0/1 -> 14, 1/1 -> 3]|[0/0 -> 0.0, 0/1 -> 0.8235294, 1/1 -> 0.1764706] |C        |0.4117647  |0/0        |0.0|\n",
      "|22:16100462:A:G|[0/0 -> 1, 0/1 -> 16]|[0/0 -> 0.05882353, 0/1 -> 0.9411765, 1/1 -> 0.0]|G        |0.47058824 |1/1        |0.0|\n",
      "|22:16105660:G:A|[0/0 -> 7, 0/1 -> 10]|[0/0 -> 0.4117647, 0/1 -> 0.5882353, 1/1 -> 0.0] |A        |0.29411766 |1/1        |0.0|\n",
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.stats(df, missingAsReference = True).selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"genotypeCount\", \"genotypeFreq\", \"mafAllele\", \"maf\", \"mgfGenotype\", \"mgf\").show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are other options we could execute:\n",
    " - Getting stats from specific individuals. In that case we need to provide a list of samples.\n",
    " - Specify the name we want to give to the stats cohort, which by default is \"ALL\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------------+------------------------------------+---------+----+-----------+---+\n",
      "|id             |genotypeCount       |genotypeFreq                        |mafAllele|maf |mgfGenotype|mgf|\n",
      "+---------------+--------------------+------------------------------------+---------+----+-----------+---+\n",
      "|22:16054454:C:T|[0/0 -> 10]         |[0/0 -> 1.0, 0/1 -> 0.0, 1/1 -> 0.0]|T        |0.0 |0/1        |0.0|\n",
      "|22:16065809:T:C|[0/0 -> 7, 0/1 -> 3]|[0/0 -> 0.7, 0/1 -> 0.3, 1/1 -> 0.0]|C        |0.15|1/1        |0.0|\n",
      "|22:16077310:T:A|[0/0 -> 5, 0/1 -> 5]|[0/0 -> 0.5, 0/1 -> 0.5, 1/1 -> 0.0]|A        |0.25|1/1        |0.0|\n",
      "|22:16080499:A:G|[0/0 -> 8, 0/1 -> 2]|[0/0 -> 0.8, 0/1 -> 0.2, 1/1 -> 0.0]|G        |0.1 |1/1        |0.0|\n",
      "|22:16084621:T:C|[0/0 -> 7, 0/1 -> 3]|[0/0 -> 0.7, 0/1 -> 0.3, 1/1 -> 0.0]|C        |0.15|1/1        |0.0|\n",
      "|22:16091610:G:T|[0/0 -> 9, 0/1 -> 1]|[0/0 -> 0.9, 0/1 -> 0.1, 1/1 -> 0.0]|T        |0.05|1/1        |0.0|\n",
      "|22:16096040:G:A|[0/0 -> 7, 0/1 -> 3]|[0/0 -> 0.7, 0/1 -> 0.3, 1/1 -> 0.0]|A        |0.15|1/1        |0.0|\n",
      "|22:16099957:C:T|[0/1 -> 10]         |[0/0 -> 0.0, 0/1 -> 1.0, 1/1 -> 0.0]|C        |0.5 |0/0        |0.0|\n",
      "|22:16100462:A:G|[0/1 -> 10]         |[0/0 -> 0.0, 0/1 -> 1.0, 1/1 -> 0.0]|A        |0.5 |0/0        |0.0|\n",
      "|22:16105660:G:A|[0/0 -> 4, 0/1 -> 6]|[0/0 -> 0.4, 0/1 -> 0.6, 1/1 -> 0.0]|A        |0.3 |1/1        |0.0|\n",
      "+---------------+--------------------+------------------------------------+---------+----+-----------+---+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "samples = oskar.metadata.samples(df)[\"hgvauser@platinum:illumina_platinum\"]\n",
    "tenSamples = list(samples[0:10])\n",
    "oskar.stats(df, studyId = \"hgvauser@platinum:illumina_platinum\", cohort = \"Caucasians\", samples = tenSamples, missingAsReference = True)\\\n",
    "    .selectExpr(\"id\", \"studies[0].stats.Caucasians as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"genotypeCount\", \"genotypeFreq\", \"mafAllele\", \"maf\", \"mgfGenotype\", \"mgf\").show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global Stats\n",
    "We could also want to use _globalStats()_ that provides statistics attached to the whole dataframe.\n",
    "<br>\n",
    "Usage:\n",
    "```python\n",
    "globalStats(df[DataFrame], studyId[str]=None, fileId[str]=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some fields will be given as *int* and *float*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "|numVariants|numSamples|numPass|tiTvRatio|meanQuality|stdDevQuality|\n",
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "|       1000|         0|   4016|1.3342036|    773.252|    765.54596|\n",
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe = oskar.globalStats(df)\n",
    "globalStatsTransformationDataframe.select(\"numVariants\", \"numSamples\", \"numPass\", \"tiTvRatio\", \"meanQuality\", \"stdDevQuality\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Others are returned as *array*, although in this case it is empty:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+\n",
      "|numRareVariants|\n",
      "+---------------+\n",
      "|             []|\n",
      "+---------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe.select(\"numRareVariants\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And others will be given as _map_. In order to show it here we previously converted it into a new DataFrame by using _explode_ PySpark command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+-----+\n",
      "|variantTypeCounts|value|\n",
      "+-----------------+-----+\n",
      "|              SNV|  894|\n",
      "|            INDEL|  106|\n",
      "+-----------------+-----+\n",
      "\n",
      "+----------------------------------+-----+\n",
      "|variantBiotypeCounts              |value|\n",
      "+----------------------------------+-----+\n",
      "|IG_C_pseudogene                   |1    |\n",
      "|IG_V_pseudogene                   |17   |\n",
      "|protein_coding                    |500  |\n",
      "|rRNA                              |4    |\n",
      "|retained_intron                   |192  |\n",
      "|nonsense_mediated_decay           |166  |\n",
      "|IG_V_gene                         |18   |\n",
      "|snRNA                             |6    |\n",
      "|snoRNA                            |6    |\n",
      "|transcribed_processed_pseudogene  |4    |\n",
      "|unitary_pseudogene                |2    |\n",
      "|IG_J_gene                         |4    |\n",
      "|transcribed_unprocessed_pseudogene|40   |\n",
      "|pseudogene                        |1    |\n",
      "|IG_C_gene                         |4    |\n",
      "|sense_intronic                    |10   |\n",
      "|processed_pseudogene              |74   |\n",
      "|unprocessed_pseudogene            |72   |\n",
      "|miRNA                             |12   |\n",
      "|lincRNA                           |102  |\n",
      "+----------------------------------+-----+\n",
      "only showing top 20 rows\n",
      "\n",
      "+----------------------------------+-----+\n",
      "|consequenceTypesCounts            |value|\n",
      "+----------------------------------+-----+\n",
      "|missense_variant                  |4    |\n",
      "|intergenic_variant                |222  |\n",
      "|2KB_downstream_variant            |146  |\n",
      "|3_prime_UTR_variant               |9    |\n",
      "|regulatory_region_variant         |764  |\n",
      "|downstream_gene_variant           |163  |\n",
      "|intron_variant                    |543  |\n",
      "|splice_region_variant             |2    |\n",
      "|upstream_gene_variant             |200  |\n",
      "|TF_binding_site_variant           |108  |\n",
      "|5_prime_UTR_variant               |1    |\n",
      "|non_coding_transcript_exon_variant|45   |\n",
      "|non_coding_transcript_variant     |385  |\n",
      "|2KB_upstream_variant              |146  |\n",
      "|splice_donor_variant              |1    |\n",
      "|NMD_transcript_variant            |140  |\n",
      "|synonymous_variant                |6    |\n",
      "+----------------------------------+-----+\n",
      "\n",
      "+---------------+-----------+\n",
      "|chromosomeStats|value      |\n",
      "+---------------+-----------+\n",
      "|22             |[1000, 0.0]|\n",
      "+---------------+-----------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe = oskar.globalStats(df)\n",
    "globalStatsTransformationDataframe.select(explode(\"variantTypeCounts\").alias(\"variantTypeCounts\", \"value\")).show()\n",
    "globalStatsTransformationDataframe.select(explode(\"variantBiotypeCounts\").alias(\"variantBiotypeCounts\", \"value\")).show(truncate = False)\n",
    "globalStatsTransformationDataframe.select(explode(\"consequenceTypesCounts\").alias(\"consequenceTypesCounts\", \"value\")).show(truncate = False)\n",
    "globalStatsTransformationDataframe.select(explode(\"chromosomeStats\").alias(\"chromosomeStats\", \"value\")).show(truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Histogram\n",
    "This next example shows how easily we can complement PyOskar API functionality with some well known python libraries like _pandas_ or _plotly_. We would like to know the frequency with which we can find variants depending on the region of our genome. For that task we will use _histogram_ PyOskar transformation, and after it we will store our transformed dataframe into a Pandas DataFrame with the default _toPandas()_ PySpark method so we can plot it later.\n",
    "<br>\n",
    "Usage:\n",
    "```python\n",
    "histogram(df[DataFrame], inputCol[str], step[int]=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+-----+\n",
      "|   start|count|\n",
      "+--------+-----+\n",
      "|16000000|  101|\n",
      "|17000000|  100|\n",
      "|18000000|   85|\n",
      "|19000000|   53|\n",
      "|20000000|   46|\n",
      "|21000000|   85|\n",
      "|22000000|   52|\n",
      "|23000000|   49|\n",
      "|24000000|   44|\n",
      "|25000000|   62|\n",
      "|26000000|   42|\n",
      "|27000000|   54|\n",
      "|28000000|   21|\n",
      "|29000000|   29|\n",
      "|30000000|   38|\n",
      "|31000000|   46|\n",
      "|32000000|   44|\n",
      "|33000000|   34|\n",
      "|34000000|   15|\n",
      "+--------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.histogram(df,\"start\",1000000).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script><script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window._Plotly) {require(['plotly'],function(plotly) {window._Plotly=plotly;});}</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script><script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window._Plotly) {require(['plotly'],function(plotly) {window._Plotly=plotly;});}</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "linkText": "Export to plot.ly",
        "plotlyServerURL": "https://plot.ly",
        "showLink": true
       },
       "data": [
        {
         "type": "bar",
         "uid": "f89b8c2d-443e-42ca-90b7-b2555e911f9d",
         "x": [
          16000000,
          17000000,
          18000000,
          19000000,
          20000000,
          21000000,
          22000000,
          23000000,
          24000000,
          25000000,
          26000000,
          27000000,
          28000000,
          29000000,
          30000000,
          31000000,
          32000000,
          33000000,
          34000000
         ],
         "y": [
          101,
          100,
          85,
          53,
          46,
          85,
          52,
          49,
          44,
          62,
          42,
          54,
          21,
          29,
          38,
          46,
          44,
          34,
          15
         ]
        }
       ],
       "layout": {
        "autosize": true,
        "xaxis": {
         "autorange": true,
         "range": [
          15500000,
          34500000
         ],
         "type": "linear"
        },
        "yaxis": {
         "autorange": true,
         "range": [
          0,
          106.3157894736842
         ],
         "type": "linear"
        }
       }
      },
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABbwAAAFoCAYAAACcxj1kAAAgAElEQVR4nO3d33Oe9X3n//1P9mAPONiz9GSPtjnQsL6FSeuRlqbATIhJh6k9nXjIDE5HuMhx4jFmoYF8vxgltDS4iZOQhCZbucRraAGnEWEAAwFM5TqJCUn5EdkGIwdh+70HWRTfvG5b8oWu3HDp8Zh5H9z3Jd9z6cPnFpeeki//pwIAAAAAgA74T8M+AQAAAAAAWAmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCYI3AAAAAACdIHgDAAAAANAJgjcAAAAAAJ0geAMAAAAA0AmCNwAAAAAAnSB4AwAAAADQCasieL/zzjt122231R133BHHjh8/Xlu2bKnR0dEaHx+vPXv2XNRxAAAAAAA+GDofvH/4wx/WlVdeWZdddtnA4L1t27aanJysubm5OnToUI2NjdXMzMyyjwMAAAAA8MHQ+eD9rrvuuiuC98LCQvV6vTp8+PDic1NTU7Vjx45lHQcAAAAA4INjVQfvo0eP1sjISJ06dWrxuenp6dq4ceOyjl/IS3Pz9Z83/2Mn55Ib967QfxUAAAAAgJWzqoP37OxsjYyM1NmzZxef279/f61fv35Zx6uqTpw4MXCe//krQw/TbQbv833eF5qXXvl17X/2aCfnRy++3GhNjDHGGGOMMcYYY4w5d3h/VnXwfvc3uBcWFhafm56erg0bNizreFXV6dOnB87PXntz6GG6zeB9vs/7QvMvh7r7Q4DL7ni00ZoYY4wxxhhjjDHGGHPu8P6s6uC9sLBQa9as6btH965du2r79u3LOn4hbmmSHvm314Z+7m0GbwAAAABguFZ18K6qmpycrK1bt9axY8dqdna2xsfH68CBA8s+fj6CdxK8AQAAAIA2dT54P/TQQ7Vu3boaHR2tXq9X69atq4cffnjx+NzcXE1MTFSv16uxsbHavXt3359f6vj5CN5J8AYAAAAA2tT54D0sgncSvAEAAACANgneLRG8k+ANAAAAALRJ8G6J4J0EbwAAAACgTYJ3SwTvJHgDAAAAAG0SvFsieCfBGwAAAABok+DdEsE7Cd4AAAAAQJsE75YI3knwBgAAAADaJHi3RPBOgjcAAAAA0CbBuyWCdxK8AQAAAIA2Cd4tEbyT4A0AAAAAtEnwbongnQRvAAAAAKBNgndLBO8keAMAAAAAbRK8WyJ4J8EbAAAAAGiT4N0SwTsJ3gAAAABAmwTvlgjeSfAGAAAAANokeLdE8E6CNwAAAADQJsG7JYJ3ErwBAAAAgDYJ3i0RvJPgDQAAAAC0SfBuieCdBG8AAAAAoE2Cd0sE7yR4AwAAAABtErxbIngnwRsAAAAAaJPg3RLBOwne6W8P/HTo597WfPobTzVaE1iOP9i2b+h7vK059Ks3hr28AAAA8KEleLdE8E6CdxK8oRnBGwAAABhE8G6J4J0E7yR4QzOCNwAAADCI4N0SwTsJ3knwhmYEbwAAAGCQVR+8X3rppdq8eXNdccUV9fGPf7zuuOOOWlhYWDx+/Pjx2rJlS42Ojtb4+Hjt2bNnea8reAfBOwne0IzgDQAAAAyy6oP3tddeW/fdd19VVZ08ebI2bdpU99577+Lxbdu21eTkZM3NzdWhQ4dqbGysZmZmlnxdwTsJ3knwhmYEbwAAAGCQVR28z5w5U5deemkdPHhw8bm77rqrdu7cWVVVCwsL1ev16vDhw4vHp6amaseOHUu+tuCdBO8keEMzgjcAAAAwyKoO3lW/DdhjY2P1/e9/vw4fPlzr169fDNxHjx6tkZGROnXq1OLHT09P18aNG5d8XcE7Cd5J8IZmBG8AAABgkFUfvJ955pm67rrraseOHbV27dqanJys+fn5qqqanZ2tkZGROnv27OLH79+/v9avX7/4eG5ubuD85Ke/Gno0aWsuuXHveT/vC830E0eGfu5tzZq//pdGa/L/7Xtu6Ofe1vz5vY81WhNjljMf+dwPhr7H25ofv/iLoa+vMcYYY4wxxpjhDe/Pqg7eJ06cqI997GP18ssvV1XVG2+8UTfccENNTk5W1e9+w/vcf8Ryenq6NmzYsPj47NmzA+for98aejRpay65ce95P+8LzcMvvjr0c29rLrvj0UZr8jePdveHAJ/+xlON1sR0Y9rmN7wBALph2NetxhjzQRzen1UdvJ999tlau3Zt33P79++vq6++uqp+ew/vNWvW9N3De9euXbV9+/YlX9stTZJbmiS3NIFmBG8AAABgkFUdvE+ePFnr1q2r++67r86ePVvz8/M1MTFRt99+++LHTE5O1tatW+vYsWM1Oztb4+PjdeDAgSVfW/BOgncSvKEZwRsAAAAYZFUH76qqF154oTZt2lTj4+N15ZVX1p133lm/+c1vFo/Pzc3VxMRE9Xq9Ghsbq927dy/rdQXvJHgnwRuaEbwBAACAQVZ98G6L4J0E7yR4QzOCNwAAADCI4N0SwTsJ3knwhmYEbwAAAGAQwbslgncSvJPgDc0I3gAAAMAggndLBO8keCfBG5oRvAEAAIBBBO+WCN5J8E6CNzQjeAMAAACDCN4tEbyT4J0Eb2hG8AYAAAAGEbxbIngnwTsJ3tCM4A0AAAAMIni3RPBOgncSvKEZwRsAAAAYRPBuieCdBO8keEMzgjcAAAAwiODdEsE7Cd5J8IZmBG8AAABgEMG7JYJ3EryT4A3NCN4AAADAIIJ3SwTvJHgnwRuaEbwBAACAQQTvlgjeSfBOgjc0I3gDAAAAgwjeLRG8k+CdBG9oRvAGAAAABhG8WyJ4J8E7Cd7QjOANAAAADCJ4t0TwToJ3EryhGcEbAAAAGETwbongnQTvJHhDM4I3AAAAMIjg3RLBOwneSfCGZgRvAAAAYBDBuyWCdxK8k+ANzQjeAAAAwCCCd0sE7yR4J8EbmhG8AQAAgEEE75YI3knwToI3NCN4AwAAAIMI3i0RvJPgnQRvaEbwBgAAAAYRvFsieCfBOwne0IzgDQAAAAwieLdE8E6CdxK8oRnBGwAAABhE8G6J4J0E7yR4QzOCNwAAADCI4F1Vzz//fG3evLkuv/zy6vV69dprry0eO378eG3ZsqVGR0drfHy89uzZs6zXFLyT4J0Eb2hG8AYAAAAGWfXB+7nnnquxsbGanp6uV199tV555ZVaWFhYPL5t27aanJysubm5OnToUI2NjdXMzMySryt4J8E7Cd7QjOANrITte18Y+nu+rdm+94VhLy8AAAzFqg/emzZtqr17BwfchYWF6vV6dfjw4cXnpqamaseOHUu+ruCdBO8keEMzgjewEgRvAADonlUdvE+ePFkjIyO1bdu2Gh8fr7Vr19bWrVvr5MmTVVV19OjRGhkZqVOnTi3+menp6dq4cePi47Nnzw6co79+a+jf6LQ1l9y497yf94Xm4RdfHfq5tzWX3fFoozX5m0ePDP3c25pPf+OpRmtizHKmy8H7hV+eGPr6GrNapuvBe9jra4wxxhhjmg3vz6oO3keOHKmRkZGamZmp+fn5eu211+ozn/lMbdu2raqqZmdna2RkpG+j7d+/v9avX7/4+NixYwPnuZ/9x9C/0WlrLrlx73k/7wvN3ie7+9vMa/76Xxqtyf//f54f+rm3NX9+72ON1sSY5cxHPveDoe/xtubHL/5i6OtrzGqZm+4/OPT3fFtz0/0Hh76+xhhjjDGm2fD+CN4jI3XmzJnF5x5//PG6/PLLq+p3v+F97j29p6ena8OGDUu+tluaJLc0SW5pAs10+Te83dIEfn+6/hveAACwGq3q4D0/P1+XXnppHTlyZPG5mZmZuuqqq6rqt/fwXrNmTd89vHft2lXbt29f8rUF7yR4J8EbmhG8gZUgeAMAQPes6uBdVfWFL3yhrr/++pqbm6vXX3+9Nm3aVPfcc8/i8cnJydq6dWsdO3asZmdna3x8vA4cOLDk6wreSfBOgjc0I3gDK0HwBgCA7ln1wfutt96qW265pdatW1dXXHFF3X333fXOO+8sHp+bm6uJiYnq9Xo1NjZWu3fvXtbrCt5J8E6CNzQjeAMrQfAGAIDuWfXBuy2CdxK8k+ANzQjewEoQvAEAoHsE75YI3knwToI3NCN4AytB8AYAgO4RvFsieCfBOwne0IzgDawEwRsAALpH8G6J4J0E7yR4QzOCN7ASBG8AAOgewbslgncSvJPgDc0I3sBKELwBAKB7BO+WCN5J8E6CNzQjeAMrQfBmOe754U/rtn0vdnJ+efzUsJcXAGDFCd4tEbyT4J0Eb2hG8AZWguDNcnz0ln8e+n/PtuaJnx8b9vICAKw4wbslgncSvJPgnb7yyJH6k6kfdXK+++QvGq0JSfAGVoLgzXII3gAAHy6Cd0sE7yR4J8E7Tdz/7NDPva350oOzjdaEJHgDK0HwZjkEbwCADxfBuyWCdxK8k+CdBG+WQ/Dud+LUO0M/7zbn9JmzLewiELxZHsEbAODDRfBuieCdBO8keCfBm+UQvPsJ3tCM4M1yCN4AAB8ugndLBO8keCfBOwneLIfg3U/whmYEb5ZD8AYA+HARvFsieCfBOwneSfBmOQTvfoI3NCN4sxyCNwDAh4vg3RLBOwneSfBOgjfLIXj3E7yhGcGb5RC8AQA+XATvlgjeSfBOgncSvFkOwbuf4J2e+PmxoZ93W/PRW/75oteDwQRvlkPwBgD4cBG8WyJ4J8E7Cd5J8GY5BO9+gncSvFkOwZvlELwBAD5cBO+WCN5J8E6CdxK8WQ7Bu5/gnQRvlkPwZjkEbwCADxfBuyWCdxK8k+CdBG+WQ/DuJ3gnwZvlELxZDsEbAODDRfBuieCdBO8keCfBm+UQvPsJ3knwZjkEb5ZD8AYA+HARvFsieCfBOwneSfBOfzL1o6Gfe1sz/cwvG62J4N1P8E6CN8sheLMcgjewEr704OzQ3/NtzcT9zw57eQH6CN4tEbyT4J0E7yR4J8E7Cd79BO8keLMcgjfLIXgDK0HwBvj9EbxbIngnwTsJ3knwToJ3Erz7Cd5J8GY5BG+WQ/AGVoLgDfD7I3i3RPBOgncSvJPgnQTvJHj3E7yT4J1+8vKJum3fi52c7x18udGaCN4sh+ANrATBO/3Ts78a+jVEW/PkUV9fYZgE75YI3knwToJ3EryT4J0E736CdxK80zcfPzr0c29rPvXVxxutieDNcgjewEoQvNOnv/HU0M+9rfnbAz9d2Q0EXBTB+/95++236/rrr6/rr7++7/njx4/Xli1banR0tMbHx2vPnj3Lej3BOwneSfBOgncSvJPg3U/wToJ3EryT4M1yCN7AShC8k+ANtEXwrqozZ87UTTfdVBs3bozgvW3btpqcnKy5ubk6dOhQjY2N1czMzJKvKXgnwTsJ3knwToJ3Erz7Cd5J8E6CdxK8WQ7BG1gJgncSvIG2CN5Vdeutt9bU1FQ98MADfcF7YWGher1eHT58ePG5qamp2rFjx5KvKXgnwTsJ3knwToJ3Erz7Cd5J8E6CdxK806e++vjQz72t+ebjRxutieANrATBOwneQFtWffC+++676+abb66zZ89G8D569GiNjIzUqVOnFp+bnp6ujRs3Lvm6gncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8AZWguCdBG+gLas6eO/bt68mJibq9OnTVVURvGdnZ2tkZKTOnv3dN9r79++v9evXLz5+5513Bs5PX31j6F9g25pLbtx73s/7QvPPL/zH0M+9rRm9/ZFaWFi46PnKw4eHfu5tzV98/YlGa/KX33l66Ofe1nxx36FGa3LFXf869HNva77/1EuNvp50OXj/5KW5i16PX795aujn3eb85u2Fi16THx95fejn3dZ89JZ/bvS++fpMd3/Ieu3f/bjRmnz+H58b+rm3NZ//x+carcm1f/fjoZ97W/P1mZ82WpMuB+8fH3m90ZoYs5LT5Pr4wzhf3Hdo6O/5tuYvv/N0ozX5i68/MfRzb2u+8vDhoe+5D8MM++vPB3l4f1Z18P7yl79ca9asqV6vV71ery699NIaGRmpXq9Xp06dWvwN74WFhcU/Mz09XRs2bFh8fOLEiYHz/M9fGfoX2Lbmkhv3nvfzvtA8cPDnQz/3tqb3xYfrjTfeuOjZ9WB3L3o27P5xozW54VtPDv3c25r/tfcnjdZk/M4DQz/3tubbj/17ozX5yOd+MPRzb2ueOPyri16PX7w6N/TzbnOOHT9x0Wvy6AsvD/2825r/fvODjd43X33034Z+7m3NNX/zo0ZrMvkP3f0h6+Q/PN1oTa75m+7+raKvPvpvjdbkv9/84NDPva159IWXG62JMebi53/t/cnQ3/NtzQ3ferLRmmzY3d0fsu568NDQ99yHYZq0pdUyvD+rOni/16B7eK9Zs6bvHt67du2q7du3L/labmmS3NIkuaVJckuT5JYmqcu/4e2WJjluadI/bmmS45YmOW5pkuOWJjluaQK/P25pktzSBGiL4H2O9wbvqqrJycnaunVrHTt2rGZnZ2t8fLwOHDiw5GsJ3knwToJ3EryT4J0E736CdxK8k+CdBO8keCfBG1gJgncSvIG2CN7nGBS85+bmamJionq9Xo2NjdXu3buX9VqCdxK8k+CdBO8keCfBu5/gnQTvJHgnwTsJ3knwhov3+sm361///fVOzk9ebnarAcE7Cd5AWwTvlgjeSfBOgncSvJPgnQTvfoJ3EryT4J0E7yR4J8EbLt70M78c+v5ua/5k6keN1kTwToI30BbBuyWCdxK8k+CdBO8keCfBu5/gnQTvJHgnwTsJ3knwhosneCfBOwneQFsE75YI3knwToJ3EryT4J0E736CdxK8k+CdBO8keCfBGy6e4J0E7yR4A20RvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb7h4gncSvJPgDbRF8G6J4J0E7yR4J8E7Cd5J8O4neCfBOwneSfBOgncSvOHiCd5J8E6CN9AWwbslgncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8IaLJ3gnwTsJ3kBbBO+WCN5J8E6CdxK8k+CdBO9+gncSvJPgnQTvJHgnwRsunuCdBO8keKevzfy8btv3Yifn579+q9GaQBOCd0sE7yR4J8E7Cd5J8E6Cdz/BOwneSfBOgncSvJPgDRdP8E6CdxK802V3PDr0c29rHvm31xqtCTQheLdE8E6CdxK8k+CdBO8kePcTvJPgnQTvJHgnwTsJ3nDxBO8keCfBOwnesDIE75YI3knwToJ3EryT4J0E736CdxK8k+CdBO8keCfBGy6e4J0E7yR4J8EbVobg3RLBOwneSfBOgncSvJPg3U/wToJ3EryT4J0E7yR4w8UTvJPgnQTvJHjDyhC8WyJ4J8E7Cd5J8E6CdxK8+wneSfBOgncSvJPgnQRvuHiCdxK8k+CdBG9YGYJ3SwTvJHgnwTsJ3knwToJ3P8E7Cd5J8E6CdxK8k+ANF0/wToJ3EryT4A0rQ/BuieCdBO8keCfBOwneSfDuJ3gnwTsJ3knwToJ3Erzh4gneSfBOgncSvGFlCN4tEbyT4J0E7yR4J8E7Cd79BO8keCfBOwneSfBOgjdcPME7Cd5J8E6CN6wMwbslgncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8GYph371xtD/W7Y1f7BtX6M1EbyT4J0E7yR4w8oQvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb5YieCfBOwneSfBOgjesDMG7JYJ3EryT4J0E7yR4J8G7n+CdBO8keCfBOwneSfBmKYJ3EryT4J0E7yR4w8oQvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb5YieCfBOwneSfBOgjesDMG7JYJ3EryT4J0E7yR4J8G7n+CdBO8keCfBOwneSfBmKYJ3EryT4J0E7yR4w8oQvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb5YieCfBOwneSfBOgjesDMG7JYJ3EryT4J0E7yR4J8G7n+CdBO8keCfBOwneSfBmKYJ3EryT4J0E7yR4w8oQvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb5YieCfBOwneSfBOgjesDMG7JYJ3EryT4J0E7yR4J8G7n+CdBO8keCfBOwneSfBmKYJ3EryT4J0E7yR4w8oQvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb5YieCfBOwneSfBOgjesjFUfvO+55576xCc+Ub1er66++ur67ne/23f8+PHjtWXLlhodHa3x8fHas2fPsl5X8E6CdxK8k+CdBO8kePcTvJPgnQTvJHgnwTsJ3ixF8E6CdxK8k+CdBG9YGas+eE9NTdXBgwfr2LFj9dhjj9Xo6Gg9+eSTi8e3bdtWk5OTNTc3V4cOHaqxsbGamZlZ8nUF7yR4J8E7Cd5J8E6Cdz/BOwneSfBOgncSvJPgzVIE7yR4J8E7Cd5J8IaVseqD93tt2rSpvv3tb1dV1cLCQvV6vTp8+PDi8ampqdqxY8eSryN4J8E7Cd5J8E6CdxK8+wneSfBOgncSvJPgnQRvliJ4J8E7Cd5J8E6CN6wMwfscp06dqnXr1tUTTzxRVVVHjx6tkZGROnXq1OLHTE9P18aNG5d8LcE7Cd5J8E6CdxK8k+DdT/BOgncSvJPgnQTvJHin2/a92NlpQvBOgncSvJPgnQRvWBmC9zl27txZmzdvrrNnf/uN9ezsbI2MjCw+rqrav39/rV+/fvHx22+/PXD+/T+OD/2LSVtzyY17z/t5X2gefK67Fz2jtz/SaE2+8vDhoZ97W/MXX3+i0Zr85XeeHvq5tzVf3Heo0Zpccde/Dv3c25rvPXm00Zp0OXg/e/TXF70er514a+jn3ebMn/rNRa/JzOFXh37ebc0f7nyo0fvmaz86MvRzb2vW3/NYozXZ9r9/MvRzb2u2/e+fNFqT9fc8NvRzb2u+9qMjjdbkD3c+NPRzb2tmDr960esxf+o3Qz/vNue1E29d9Jo8e/TXQz/vtuYPtu1r9L753pPd/SHrFXf9a6M1+eK+Q0M/97bmL7/zdKM1+YuvPzH0c29rvvLw4UZrMnr7I0M/97bmwed+2WhNVuvw/gje/8+dd95Z1113XZ08eXLxuXd/w3thYWHxuenp6dqwYcPi45MnTw6cF3/x+tC/mLQ1l9y497yf94Vm3zMvDf3c25reFx9utCZ3PfTi0M+9rdn49483WpPN93X3p/y3/tNzjdZk/M4DQz/3tuY7Pz7SaE0+8rkfDP3c25qnjrxy0evxy9e7+0PW/7z5H+vEG29e9JocONTdH7L+4c6HGr1v7j3Q3d8s++TfzjRak63fe2bo597WbP3eM43W5JN/OzP0c29r7j0w22hNuhy8Dxz65UWvx4k33hz6ebc5v3z9+EWvyVNHXhn6ebc1H/ncDxq9b77z4+7+kHX8zgON1uTWf3pu6Ofe1my+76lGa7Lx77v7t4rueujFRmvS++LDQz/3tmbfMy81WpPVOrw/qz54nz59unbu3FmbNm2KDbWwsFBr1qzpu4f3rl27avv27Uu+rluaJLc0SW5pktzSJLmlSeryb3i7pUmOW5r0j1ua5LilSY5bmuS4pUlOk1uanD5zdujn3eacOPXORa+JW5oktzRJbmmS3NIkuaUJrIxVHbxPnz5dN9xwQ9144411/Pjxmp+fr/n5+b57dk9OTtbWrVvr2LFjNTs7W+Pj43XgwIElX1vwToJ3EryT4J0E7yR49xO8k+CdBO8keCfBOwne/QTvJHgnwTsJ3knwToI3rIxVHbzffPPNGhkZibnmmmsWP2Zubq4mJiaq1+vV2NhY7d69e1mvLXgnwTsJ3knwToJ3Erz7Cd5J8E6CdxK8k+CdBO9+gncSvJPgnQTvJHgnwRtWxkpXzFgAABFBSURBVKoO3m0SvJPgnQTvJHgnwTsJ3v0E7yR4J8E7Cd5J8E6Cdz/BOwneSfBOgncSvJPgDStD8G6J4J0E7yR4J8E7Cd5J8O4neCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0Eb1gZgndLBO8keCfBOwneSfBOgnc/wTsJ3knwToJ3EryT4N1P8E6CdxK8k+CdBO8keMPKELxbIngnwTsJ3knwToJ3Erz7Cd5J8E6CdxK8k+CdBO9+gncSvJPgnQTvJHgnwRtWhuDdEsE7Cd5J8E6CdxK8k+DdT/BOgncSvJPgnQTvJHj3E7yT4J0E7yR4J8E7Cd7pSw/O1m37XuzkNPl/DssjeLdE8E6CdxK8k+CdBO8kePcTvJPgnQTvJHgnwTsJ3v0E7yR4J8E7Cd5J8E6Cd7rkxr1DP/e25qW5+UZrwtIE75YI3knwToJ3EryT4J0E736CdxK8k+CdBO8keCfBu5/gnQTvJHgnwTsJ3knwToI3TQjeLRG8k+CdBO8keCfBOwne/QTvJHgnwTsJ3knwToJ3P8E7Cd5J8E6CdxK8k+CdBG+aELxbIngnwTsJ3knwToJ3Erz7Cd5J8E6CdxK8k+CdBO9+gncSvJPgnQTvJHgnwTsJ3jQheLdE8E6CdxK8k+CdBO8kePcTvJPgnQTvJHgnwTsJ3v0E7yR4J8E7Cd5J8E6CdxK8aULwbongnQTvJHgnwTsJ3knw7id4J8E7Cd5J8E6CdxK8+wneSfBOgncSvJPgnQTvJHjThODdEsE7Cd5J8E6CdxK8k+DdT/BOgncSvJPgnQTvJHj3E7yT4J0E7yR4J8E7Cd5J8KYJwbslgncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8O4neCfBOwneSfBOgncSvJPgTROCd0sE7yR4J8E7Cd5J8E6Cdz/BOwneSfBOgncSvJPg3U/wToJ3EryT4J0E7yR4J8GbJgTvlgjeSfBOgncSvJPgnQTvfoJ3EryT4J0E7yR4J8G7n+CdBO8keCfBOwneSfBOgjdNCN4tEbyT4J0E7yR4J8E7Cd79BO8keCfBOwneSfBOgnc/wTsJ3knwToJ3EryT4J0Eb5oQvFsieCfBOwneSfBOgncSvPsJ3knwToJ3EryT4J0E736CdxK8k+CdBO8keCfBOwneNCF4t0TwToJ3EryT4J0E7yR49xO8k+CdBO8keCfBOwne/QTvJHgnwTsJ3knwToJ3ErxpQvBuieCdBO8keCfBOwneSfDuJ3gnwTsJ3knwToJ3Erz7Cd5J8E6CdxK8k+CdBO8keNOE4N0SwTsJ3knwToJ3EryT4N1P8E6CdxK8k+CdBO8kePcTvJPgnQTvJHgnwTsJ3knwpgnBuyWCdxK8k+CdBO8keCfBu5/gnQTvJHgnwTsJ3knw7id4J8E7Cd5J8E6CdxK8k+BNE4J3SwTvJHgnwTsJ3knwToJ3P8E7Cd5J8E6CdxK8k+DdT/BOgncSvJPgnQTvJHgnwZsmBO+WCN5J8E6CdxK8k+CdBO9+gncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8E6CN00I3i0RvJPgnQTvJHgnwTsJ3v0E7yR4J8E7Cd5J8E6Cdz/BOwneSfBOgncSvJPgnQRvmhC8WyJ4J8E7Cd5J8E6CdxK8+wneSfBOgncSvJPgnQTvfoJ3EryT4J0E7yR4J8E7Cd40IXgv4fjx47Vly5YaHR2t8fHx2rNnz7L+nOCdBO8keCfBOwneSfDuJ3gnwTsJ3knwToJ3Erz7Cd5J8E6CdxK8k+CdBO8keNOE4L2Ebdu21eTkZM3NzdWhQ4dqbGysZmZmlvxzgncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8O4neCfBOwneSfBOgncSvJPgTROC9wUsLCxUr9erw4cPLz43NTVVO3bsWPLPCt5J8E6CdxK8k+CdBO9+gncSvJPgnQTvJHgnwbuf4J0E7yR4J8E7Cd5J8E6CN00I3hdw9OjRGhkZqVOnTi0+Nz09XRs3blzyzwreSfBOgncSvJPgnQTvfoJ3EryT4J0E7yR4J8G7n+CdBO8keCfBOwneSfBOgjdNCN4XMDs7WyMjI3X27O++0d6/f3+tX79+8fG99947cG6/+++H/sZpa/7LZ79/3s/7QvNXd+4Z+rm3Nf/tpn9otCZ//tffGvq5tzWXb/9OozX5nzd/Z+jn3tZce+t9jdbko5/7h6Gfe1tzw5f2NFqT/zrx/aGfe1tzy5e/dtHrMXXP7qGfd5vzd1+9+D3y+bu+PvTzbms+suV7jd43n779G0M/97bmf3z+u43W5Kpbvj30c29rrrrl243W5H98/rtDP/e25tO3f6PRmnxky/eGfu5tzefv+vpFr8ffffXeoZ93mzN1z+6LXpNbvvy1oZ93W/NfJ5p973fDl7r7vd9HP9fse79rb71v6Ofe1vzPm5t973f59u5+7/fnf/2tRmvy327q7vd+f3Vns+/9/stnu/u93+13//15P2/eH8H7At79De+FhYXF56anp2vDhg2Lj7/61a8aY4wxxhhjjDHGGGPMigzvj+B9AQsLC7VmzZq+e3jv2rWrtm/fPsSzAgAAAABgEMF7CZOTk7V169Y6duxYzc7O1vj4eB04cGDYpwUAAAAAwHsI3kuYm5uriYmJ6vV6NTY2Vrt37x72KQEAAAAAMIDg/QH2zjvv1G233VZ33HHHwOPPP/98bd68uS6//PLq9Xr12mu/+xdvr7322vrjP/7jmp//3b/4evr06bryyivrj/7oj1o/d35/LrRPXnrppdq8eXNdccUV9fGPf7zuuOOOvnvS2yerwz333FOf+MQnqtfr1dVXX13f/e53+44fP368tmzZUqOjozU+Pl579uzpO26fdN9Se2Sp4/bI6rDUPnjX22+/Xddff31df/31fc/bJ6vDcvaJa1iW2ieuYfnWt75Vn/rUp2p0dLTWrVtXN998c7311luLx12/UrX0PnENS9XS++RdrmG7RfD+gPrhD39YV155ZV122WUDQ+Zzzz1XY2NjNT09Xa+++mq98sorcRF41VVX1X333bf43IMPPlhXX321N2SHLLVPrr322sU9cPLkydq0aVPfv/Zrn6wOU1NTdfDgwTp27Fg99thjNTo6Wk8++eTi8W3bttXk5GTNzc3VoUOHamxsrGZmZhaP2yfdt9QeWeq4PbI6LLUPqqrOnDlTN910U23cuHHgNwv2SfcttU9cw1K1vP+vuIZd3R544IF64YUX6sSJE/Wzn/2sPvnJT9Y3v/nNxeOuX6laep+4hqVq6X1S5Rq2iwTvD7i77rprYMjctGlT7d2797x/7tprr63777+//vRP/7TeeeedqqrasGFD3X///d6QHTRon5w5c6YuvfTSOnjwYN/H7dy5c/GxfbI6bdq0qb797W9X1W//cd5er9f3j/NOTU3Vjh07Fh/bJ6vPuXtkOcftkdVp0D659dZba2pqqh544IGB3yzYJ6vPe/eJa1gGOXefuIblXGfOnKkjR47UNddcU0899VRVuX4lDdong7iGXd0utE9cw3aP4P0BNyhknjx5skZGRmrbtm01Pj5ea9eura1bt9bJkycXP+baa6+tgwcP1mc/+9l64IEH6umnn67PfOYz9fzzz3tDdtD5fjAyNTVVY2Nj9f3vf78OHz5c69ev77swtE9Wn1OnTtW6devqiSeeqKqqo0eP1sjISJ06dWrxY6anp2vjxo2Lj+2T1eW9e2Q5x+2R1WfQPrj77rvr5ptvrrNnz573mwX7ZHV57z5xDcsgg76euIblXSMjI7VmzZq+H5S5fuW9Bu2T93INy/n2iWvYbhK8P+AGhcwjR47UyMhIzczM1Pz8fL322mv1mc98prZt27b4Me++IZ988slav359TUxM1GOPPeYN2VHnC97PPPNMXXfddbVjx45au3ZtTU5O9t13yj5ZfXbu3FmbN2+us2fPVlXV7OxsjYyMLD6uqtq/f3+tX79+8bF9srq8d48s57g9svq8dx/s27evJiYm6vTp01VVF/xmwT5ZPd67T1zDMsig/6+4huVd7/5G5lVXXVWPPPJIVbl+JQ3aJ+/lGpZB+8Q1bHcJ3h9wFwreZ86cWXzu8ccfr8svv3zx8btvyKrf/lWLP/uzP6uq8obsqEH75MSJE/Wxj32sXn755aqqeuONN+qGG26oycnJxY+xT1aXO++8s6677rq+36R79zdkzr1/6vT0dG3YsGHxsX2yegzaI8s5bo+sLoP2wZe//OVas2ZN9Xq96vV6demll9bIyEj1er3F38CzT1aXQfvENSzvNWifuIZlkK985SuLPxxz/cr5nLtPzuUalnOdu09cw3aX4P0BNyhkzs/P16WXXlpHjhxZfG5mZqauuuqqxcfnviGfffbZevrpp6vKG7KrBu2TZ599ttauXdv33P79++vqq69efGyfrA6nT5+unTt31qZNm+Iib2FhodasWdP314R37dpV27dvX3xsn3TfhfbIco7bI6vDUvvgXBf67Zgq+6TLLrRPXMPyrgvtE9ewDHL77bfXzTffXFWuXzm/c/dJlWtYBnvvPjmXa9juELw/4M53q4ovfOELdf3119fc3Fy9/vrrtWnTprrnnnsWj5/7hjyXN2Q3ne9e7+vWrav77ruvzp49W/Pz8zUxMVG333774sfYJ913+vTpuuGGG+rGG2+s48eP1/z8fM3Pz/fd83BycrK2bt1ax44dq9nZ2RofH68DBw4sHrdPum2pPbKcPWSPdN9y9sG5lvpm4Vz2SXcsZ5+4hmWpfeIaljNnztRf/dVf1VNPPVVzc3P1yCOP1Nq1a+uxxx5b/BjXryy1T1zDUrW8ryfncg3bHYL3B9RDDz1U69atq9HR0er1erVu3bp6+OGHF4+/9dZbdcstt9S6devqiiuuqLvvvnvxX4ut8oZcLZbaJy+88EJt2rSpxsfH68orr6w777yzfvOb3ywet0+6780336yRkZGYa665ZvFj5ubmamJionq9Xo2NjdXu3bv7XsM+6bal9shy9pA90n3L2Qfn8s3C6rScfeIaluXsE9ewq9vZs2frlltuqauvvrp6vV6tX7++9u3b1/cxrl9Zap+4hqVqeV9PzuUatjsEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOkHwBgAAAACgEwRvAAAAAAA6QfAGAAAAAKATBG8AAAAAADpB8AYAAAAAoBMEbwAAAAAAOuH/AttJM6rSPjNJAAAAAElFTkSuQmCC",
      "text/html": [
       "<div id=\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\", [{\"x\": [16000000, 17000000, 18000000, 19000000, 20000000, 21000000, 22000000, 23000000, 24000000, 25000000, 26000000, 27000000, 28000000, 29000000, 30000000, 31000000, 32000000, 33000000, 34000000], \"y\": [101, 100, 85, 53, 46, 85, 52, 49, 44, 62, 42, 54, 21, 29, 38, 46, 44, 34, 15], \"type\": \"bar\", \"uid\": \"c5ccc0ec-e292-4337-8a95-74383d5d7d45\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\"})});</script><script type=\"text/javascript\">window.addEventListener(\"resize\", function(){window._Plotly.Plots.resize(document.getElementById(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\"));});</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<div id=\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\", [{\"x\": [16000000, 17000000, 18000000, 19000000, 20000000, 21000000, 22000000, 23000000, 24000000, 25000000, 26000000, 27000000, 28000000, 29000000, 30000000, 31000000, 32000000, 33000000, 34000000], \"y\": [101, 100, 85, 53, 46, 85, 52, 49, 44, 62, 42, 54, 21, 29, 38, 46, 44, 34, 15], \"type\": \"bar\", \"uid\": \"c5ccc0ec-e292-4337-8a95-74383d5d7d45\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\"})});</script><script type=\"text/javascript\">window.addEventListener(\"resize\", function(){window._Plotly.Plots.resize(document.getElementById(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\"));});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly.offline as py\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import get_plotlyjs, init_notebook_mode\n",
    "\n",
    "pandasHistogramDF = oskar.histogram(df,\"start\",1000000).toPandas()\n",
    "\n",
    "init_notebook_mode(connected = True)\n",
    "\n",
    "data = [go.Bar(x= pandasHistogramDF[\"start\"], y = pandasHistogramDF[\"count\"])]\n",
    "py.iplot(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** If you are using jupyter lab you will need to download this extension https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/plotly-extension to visualize the plot*"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
