{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Statistics tutorial**\n",
    "In this second tutorial we will show some examples of how to executed more sofisticated analysis towards our data. Here we will not use UDFs but transformers. These kind of functions depend directly from an Oskar object, and they will not return a dataframe field but the whole dataframe again with the due transformation applied."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we did in the previous tutorial, we need to import both Spark and Oskar APIs, as well as loading our data into a spark DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyoskar.core import Oskar\n",
    "from pyoskar.sql import *\n",
    "from pyoskar.analysis import *\n",
    "from pyspark.sql.functions import col, udf, count, explode, concat, when, expr\n",
    "from pyspark.sql.functions import *\n",
    "\n",
    "oskar = Oskar(spark)\n",
    "df = oskar.load(\"/home/roldanx/appl/oskar/oskar-spark/src/test/resources/platinum_chr22.small.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats per variant\n",
    "Here we present the genotypes of the first 10 variants from which the later stats will come out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "|id             |genotypes                                                                                                                |\n",
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "|22:16054454:C:T|[[[./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [./.], [./.], [./.]]]|\n",
      "|22:16065809:T:C|[[[0/1], [./.], [./.], [./.], [0/1], [./.], [0/1], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [./.], [./.], [0/1]]]|\n",
      "|22:16077310:T:A|[[[./.], [0/1], [0/1], [./.], [./.], [0/1], [./.], [0/1], [./.], [0/1], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16080499:A:G|[[[0/1], [./.], [0/1], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [./.], [./.], [./.]]]|\n",
      "|22:16084621:T:C|[[[./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16091610:G:T|[[[./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16096040:G:A|[[[./.], [./.], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [0/1]]]|\n",
      "|22:16099957:C:T|[[[0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [1/1], [0/1], [1/1], [1/1], [0/1], [0/1]]]|\n",
      "|22:16100462:A:G|[[[0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1]]]|\n",
      "|22:16105660:G:A|[[[0/1], [./.], [0/1], [./.], [0/1], [./.], [0/1], [0/1], [./.], [0/1], [./.], [0/1], [0/1], [./.], [0/1], [./.], [0/1]]]|\n",
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.select(df.id, df.studies.samplesData.alias(\"genotypes\")).show(10, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case we are applying a transformation that will lead to adding new data inside the \"studies\" field, which is empty before we do anything:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16065809:T:C|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16077310:T:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16080499:A:G|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16084621:T:C|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16091610:G:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16096040:G:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16099957:C:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16100462:A:G|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16105660:G:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\").show(10, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we proceed to call our [<span style=\"color:#FF1493\"> **stats** </span>] transformer through our \"oskar\" instance. That means that our dataframe will suffer some kind of change. It is important to know how transformers work in order to tell spark how to extract the new information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|2          |1             |1             |32                |16                  |0.5          |0.5          |\n",
      "|22:16065809:T:C|12         |6             |6             |22                |11                  |0.5          |0.5          |\n",
      "|22:16077310:T:A|14         |7             |7             |20                |10                  |0.5          |0.5          |\n",
      "|22:16080499:A:G|8          |4             |4             |26                |13                  |0.5          |0.5          |\n",
      "|22:16084621:T:C|10         |5             |5             |24                |12                  |0.5          |0.5          |\n",
      "|22:16091610:G:T|6          |3             |3             |28                |14                  |0.5          |0.5          |\n",
      "|22:16096040:G:A|18         |9             |9             |16                |8                   |0.5          |0.5          |\n",
      "|22:16099957:C:T|34         |14            |20            |0                 |0                   |0.4117647    |0.5882353    |\n",
      "|22:16100462:A:G|32         |16            |16            |2                 |1                   |0.5          |0.5          |\n",
      "|22:16105660:G:A|20         |10            |10            |14                |7                   |0.5          |0.5          |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "statsDF = oskar.stats(df)\n",
    "statsDF.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\").show(10, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since our data has some empty genotypes, we would like to treat those missing values as if they were the reference. For that task we dispose of \"missingAsReference\" parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|34         |33            |1             |0                 |0                   |0.9705882    |0.029411765  |\n",
      "|22:16065809:T:C|34         |28            |6             |0                 |0                   |0.8235294    |0.1764706    |\n",
      "|22:16077310:T:A|34         |27            |7             |0                 |0                   |0.7941176    |0.20588236   |\n",
      "|22:16080499:A:G|34         |30            |4             |0                 |0                   |0.88235295   |0.11764706   |\n",
      "|22:16084621:T:C|34         |29            |5             |0                 |0                   |0.85294116   |0.14705883   |\n",
      "|22:16091610:G:T|34         |31            |3             |0                 |0                   |0.9117647    |0.0882353    |\n",
      "|22:16096040:G:A|34         |25            |9             |0                 |0                   |0.7352941    |0.2647059    |\n",
      "|22:16099957:C:T|34         |14            |20            |0                 |0                   |0.4117647    |0.5882353    |\n",
      "|22:16100462:A:G|34         |18            |16            |0                 |0                   |0.5294118    |0.47058824   |\n",
      "|22:16105660:G:A|34         |24            |10            |0                 |0                   |0.7058824    |0.29411766   |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "missingAsReferenceStatsDF = oskar.stats(df, missingAsReference=True)\n",
    "missingAsReferenceStatsDF.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\").show(10, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Other statistics that our method provides are the next ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "|id             |genotypeCount        |genotypeFreq                                     |mafAllele|maf        |mgfGenotype|mgf|\n",
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "|22:16054454:C:T|[0/0 -> 16, 0/1 -> 1]|[0/0 -> 0.9411765, 0/1 -> 0.05882353, 1/1 -> 0.0]|T        |0.029411765|1/1        |0.0|\n",
      "|22:16065809:T:C|[0/0 -> 11, 0/1 -> 6]|[0/0 -> 0.64705884, 0/1 -> 0.3529412, 1/1 -> 0.0]|C        |0.1764706  |1/1        |0.0|\n",
      "|22:16077310:T:A|[0/0 -> 10, 0/1 -> 7]|[0/0 -> 0.5882353, 0/1 -> 0.4117647, 1/1 -> 0.0] |A        |0.20588236 |1/1        |0.0|\n",
      "|22:16080499:A:G|[0/0 -> 13, 0/1 -> 4]|[0/0 -> 0.7647059, 0/1 -> 0.23529412, 1/1 -> 0.0]|G        |0.11764706 |1/1        |0.0|\n",
      "|22:16084621:T:C|[0/0 -> 12, 0/1 -> 5]|[0/0 -> 0.7058824, 0/1 -> 0.29411766, 1/1 -> 0.0]|C        |0.14705883 |1/1        |0.0|\n",
      "|22:16091610:G:T|[0/0 -> 14, 0/1 -> 3]|[0/0 -> 0.8235294, 0/1 -> 0.1764706, 1/1 -> 0.0] |T        |0.0882353  |1/1        |0.0|\n",
      "|22:16096040:G:A|[0/0 -> 8, 0/1 -> 9] |[0/0 -> 0.47058824, 0/1 -> 0.5294118, 1/1 -> 0.0]|A        |0.2647059  |1/1        |0.0|\n",
      "|22:16099957:C:T|[0/1 -> 14, 1/1 -> 3]|[0/0 -> 0.0, 0/1 -> 0.8235294, 1/1 -> 0.1764706] |C        |0.4117647  |0/0        |0.0|\n",
      "|22:16100462:A:G|[0/0 -> 1, 0/1 -> 16]|[0/0 -> 0.05882353, 0/1 -> 0.9411765, 1/1 -> 0.0]|G        |0.47058824 |1/1        |0.0|\n",
      "|22:16105660:G:A|[0/0 -> 7, 0/1 -> 10]|[0/0 -> 0.4117647, 0/1 -> 0.5882353, 1/1 -> 0.0] |A        |0.29411766 |1/1        |0.0|\n",
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.stats(df, missingAsReference=True).selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"genotypeCount\", \"genotypeFreq\", \"mafAllele\", \"maf\", \"mgfGenotype\", \"mgf\").show(10, False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats per dataframe\n",
    "We could also use [<span style=\"color:#FF1493\"> **globalStats** </span>] that provides statistics attached to the whole dataframe. Some fields will be given as *int* and *float*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "|numVariants|numSamples|numPass|tiTvRatio|meanQuality|stdDevQuality|\n",
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "|       1000|         0|   4016|1.3342036|    773.252|    765.54596|\n",
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe = oskar.globalStats(df)\n",
    "globalStatsTransformationDataframe.select(\"numVariants\", \"numSamples\", \"numPass\", \"tiTvRatio\", \"meanQuality\", \"stdDevQuality\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Others are returned as *array*, although in this case it is empty:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+\n",
      "|numRareVariants|\n",
      "+---------------+\n",
      "|             []|\n",
      "+---------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe.select(globalStatsTransformationDataframe.numRareVariants).show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And others will be given as *map*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+-----+\n",
      "|variantTypeCounts|value|\n",
      "+-----------------+-----+\n",
      "|              SNV|  894|\n",
      "|            INDEL|  106|\n",
      "+-----------------+-----+\n",
      "\n",
      "+----------------------------------+-----+\n",
      "|variantBiotypeCounts              |value|\n",
      "+----------------------------------+-----+\n",
      "|IG_C_pseudogene                   |1    |\n",
      "|IG_V_pseudogene                   |17   |\n",
      "|protein_coding                    |500  |\n",
      "|rRNA                              |4    |\n",
      "|retained_intron                   |192  |\n",
      "|nonsense_mediated_decay           |166  |\n",
      "|IG_V_gene                         |18   |\n",
      "|snRNA                             |6    |\n",
      "|snoRNA                            |6    |\n",
      "|transcribed_processed_pseudogene  |4    |\n",
      "|unitary_pseudogene                |2    |\n",
      "|IG_J_gene                         |4    |\n",
      "|transcribed_unprocessed_pseudogene|40   |\n",
      "|pseudogene                        |1    |\n",
      "|IG_C_gene                         |4    |\n",
      "|sense_intronic                    |10   |\n",
      "|processed_pseudogene              |74   |\n",
      "|unprocessed_pseudogene            |72   |\n",
      "|miRNA                             |12   |\n",
      "|lincRNA                           |102  |\n",
      "+----------------------------------+-----+\n",
      "only showing top 20 rows\n",
      "\n",
      "+----------------------------------+-----+\n",
      "|consequenceTypesCounts            |value|\n",
      "+----------------------------------+-----+\n",
      "|missense_variant                  |4    |\n",
      "|intergenic_variant                |222  |\n",
      "|2KB_downstream_variant            |146  |\n",
      "|3_prime_UTR_variant               |9    |\n",
      "|regulatory_region_variant         |764  |\n",
      "|downstream_gene_variant           |163  |\n",
      "|intron_variant                    |543  |\n",
      "|splice_region_variant             |2    |\n",
      "|upstream_gene_variant             |200  |\n",
      "|TF_binding_site_variant           |108  |\n",
      "|5_prime_UTR_variant               |1    |\n",
      "|non_coding_transcript_exon_variant|45   |\n",
      "|non_coding_transcript_variant     |385  |\n",
      "|2KB_upstream_variant              |146  |\n",
      "|splice_donor_variant              |1    |\n",
      "|NMD_transcript_variant            |140  |\n",
      "|synonymous_variant                |6    |\n",
      "+----------------------------------+-----+\n",
      "\n",
      "+---------------+-----------+\n",
      "|chromosomeStats|value      |\n",
      "+---------------+-----------+\n",
      "|22             |[1000, 0.0]|\n",
      "+---------------+-----------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe = oskar.globalStats(df)\n",
    "globalStatsTransformationDataframe.select(explode(\"variantTypeCounts\").alias(\"variantTypeCounts\", \"value\")).show()\n",
    "globalStatsTransformationDataframe.select(explode(\"variantBiotypeCounts\").alias(\"variantBiotypeCounts\", \"value\")).show(truncate=False)\n",
    "globalStatsTransformationDataframe.select(explode(\"consequenceTypesCounts\").alias(\"consequenceTypesCounts\", \"value\")).show(truncate=False)\n",
    "globalStatsTransformationDataframe.select(explode(\"chromosomeStats\").alias(\"chromosomeStats\", \"value\")).show(truncate=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Histogram\n",
    "This next example shows how easily we can complement Pyoskar API functionality and at the same time we use some well known python functionality like pandas or plotly. We would like to know the frequency with which we can find variants depending on the region of our genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+-----+\n",
      "|   start|count|\n",
      "+--------+-----+\n",
      "|16000000|  101|\n",
      "|17000000|  100|\n",
      "|18000000|   85|\n",
      "|19000000|   53|\n",
      "|20000000|   46|\n",
      "|21000000|   85|\n",
      "|22000000|   52|\n",
      "|23000000|   49|\n",
      "|24000000|   44|\n",
      "|25000000|   62|\n",
      "|26000000|   42|\n",
      "|27000000|   54|\n",
      "|28000000|   21|\n",
      "|29000000|   29|\n",
      "|30000000|   38|\n",
      "|31000000|   46|\n",
      "|32000000|   44|\n",
      "|33000000|   34|\n",
      "|34000000|   15|\n",
      "+--------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.histogram(df,\"start\",1000000).show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For that we will give use to the [<span style=\"color:#FF1493\"> **histogram** </span>] Pyoskar transformation, but after it we will store our transformed dataframe into a Pandas Dataframe in order to plot it later with the default \"toPandas()\" Pyspark method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script><script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window._Plotly) {require(['plotly'],function(plotly) {window._Plotly=plotly;});}</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script><script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window._Plotly) {require(['plotly'],function(plotly) {window._Plotly=plotly;});}</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "linkText": "Export to plot.ly",
        "plotlyServerURL": "https://plot.ly",
        "showLink": true
       },
       "data": [
        {
         "type": "bar",
         "uid": "d1375bc5-1307-4417-beeb-19753d162784",
         "x": [
          16000000,
          17000000,
          18000000,
          19000000,
          20000000,
          21000000,
          22000000,
          23000000,
          24000000,
          25000000,
          26000000,
          27000000,
          28000000,
          29000000,
          30000000,
          31000000,
          32000000,
          33000000,
          34000000
         ],
         "y": [
          101,
          100,
          85,
          53,
          46,
          85,
          52,
          49,
          44,
          62,
          42,
          54,
          21,
          29,
          38,
          46,
          44,
          34,
          15
         ]
        }
       ],
       "layout": {
        "autosize": true,
        "xaxis": {
         "autorange": true,
         "range": [
          15500000,
          34500000
         ],
         "type": "linear"
        },
        "yaxis": {
         "autorange": true,
         "range": [
          0,
          106.3157894736842
         ],
         "type": "linear"
        }
       }
      },
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABcIAAAFoCAYAAAB0RohPAAAgAElEQVR4nO3d73Nd9X3g8f1POjt50ift9NE+aR7eur6KSaKVhk3Abahp17P27IzWZHBT4SBXE8eYQhbSYEScUnCzTviRpEnFEmpIoIbKQ8AGB2wi4xInNAEcyTZCAmHrsw+y3Np8vrYuBx1ffO7rNfN5cO+R7xx9/b320ZvL8X8KAAAAAABosP/U6xMAAAAAAIA6CeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKMJ4QAAAAAANJoQDgAAAABAownhAAAAAAA0mhAOAAAAAECjCeEAAAAAADSaEA4AAAAAQKP1RQh/991349Zbb43bb789HTt58mRs2bIlBgYGYnh4OPbs2fOBjgMAAAAA8NHW+BD+5JNPxlVXXRWf+MQniiF8fHw8xsbGYmZmJo4cORJDQ0MxNTXV9XEAAAAAAD7aGh/C33PnnXemEL64uBjtdjuOHj3aeW5iYiK2b9/e1XEAAAAAAD76+jqEHz9+PFqtViwsLHSem5ycjI0bN3Z1/GJ+M/dO/M7mf2rs7D/2mxX6nQEAAAAAqFdfh/Dp6elotVqxtLTUeW7v3r2xbt26ro5HRJw6dao4r/zqRM9jdZ3z2KHjF/zeLzY//ukvYu+h442co6++UWlNjGnqzM3N9fwcjDHGGGPM5TWuIY0x5sLDh9PXIfy9T3wvLi52npucnIwNGzZ0dTwi4syZM8V5/dR8z2N1nfPU9OsX/N4vNr8/9sOen3tds/upf6u0JsY0dc6ePdvzczDGGGOMMZfXuIY0xpgLDx9OX4fwxcXFWL169Xn3AN+5c2ds27atq+MX49YoZU0O4d+c+nmlNQEAAAAA6tXXITwiYmxsLLZu3Rqzs7MxPT0dw8PDsW/fvq6PX4gQXiaEAwAAAACXWuND+GOPPRaDg4MxMDAQ7XY7BgcH4/HHH+8cn5mZidHR0Wi32zE0NBS7d+8+79cvd/xChPAyIRwAAAAAuNQaH8J7RQgvE8IBAAAAgEtNCK+JEF4mhAMAAAAAl5oQXhMhvEwIBwAAAAAuNSG8JkJ4mRAOAAAAAFxqQnhNhPAyIRwAAAAAuNSE8JoI4WVCOAAAAABwqQnhNRHCy4RwAAAAAOBSE8JrIoSXCeEAAAAAwKUmhNdECC8TwgEAAACAS00Ir4kQXiaEAwAAAACXmhBeEyG8TAgHAAAAAC41IbwmQniZEA4AAAAAXGpCeE2E8DIhHAAAAAC41ITwmgjhZUI4AAAAAHCpCeE1EcLLhHAAAAAA4FITwmsihJcJ4QAAAADApSaE10QILxPCAQAAAIBLTQiviRBeJoQDAAAAAJeaEF4TIbxMCAcAAAAALjUhvCZCeJkQDgAAAABcakJ4TYTwMiEcAAAAALjUhPCaCOFlQnj2X+94sufnXtf89Q9eqLQm0I3HDr/W8z1e57z97tleLzEAAAA0hhBeEyG8TAjPhHCoRggHAAAAuiWE10QILxPCMyEcqhHCAQAAgG4J4TURwsuE8EwIh2qEcAAAAKBbfR/Cf/GLX8TmzZvjyiuvjM985jNx++23x+LiYuf4yZMnY8uWLTEwMBDDw8OxZ8+erl5XCC8TwjMhHKoRwgEAAIBu9X0Iv/baa+P++++PiIi5ubkYGRmJe++9t3N8fHw8xsbGYmZmJo4cORJDQ0MxNTW17OsK4WVCeCaEQzVCOAAAANCtvg7hZ8+ejVWrVsXBgwc7z915552xY8eOiIhYXFyMdrsdR48e7RyfmJiI7du3L/vaQniZEJ4J4VCNEA4AAAB0q69DeMRvw/bQ0FB8//vfj6NHj8a6des64fv48ePRarViYWGh8/WTk5OxcePGZV9XCC8TwjMhHKoRwgEAAIBu9X0If/7552P9+vWxffv2WLNmTYyNjcX8/HxERExPT0er1YqlpaXO1+/duzfWrVvXeTwzM1Ocl3/Z7ECz97lXLvi9X2x+78aHe37udc3XHztcaU0+efvjPT/3uuav7n+m0pqYZsybb75Z6+t//+mXe77H65xfvX6i57+HxhhjjDGXek6fPt3zczDGmI/q8OH0dQg/depUfPKTn4xXX301IiJOnz4d119/fYyNjUXEf3wi/Nx/PHNycjI2bNjQeby0tFScE2++3fOIUudMvXzigt/7xabJnwj/h399pdKaNP0T4VXWxDRn6uQT4QAAzdPr61djjPkoDx9OX4fwQ4cOxZo1a857bu/evbF27dqI+O09wlevXn3ePcJ37twZ27ZtW/a13RqlrMkh3K1R8rg1CnUSwgEAAIBu9XUIn5ubi8HBwbj//vtjaWkp5ufnY3R0NG677bbO14yNjcXWrVtjdnY2pqenY3h4OPbt27fsawvhZUJ4JoRDNUI4AAAA0K2+DuEREYcPH46RkZEYHh6Oq666Ku644454++23O8dnZmZidHQ02u12DA0Nxe7du7t6XSG8TAjPhHCoRggHAAAAutX3IbwuQniZEJ4J4VCNEA4AAAB0SwiviRBeJoRnQjhUI4QDAAAA3RLCayKElwnhmRAO1QjhAAAAQLeE8JoI4WVCeCaEQzVCOAAAANAtIbwmQniZEJ4J4VCNEA4AAAB0SwiviRBeJoRnQjhUI4QDAAAA3RLCayKElwnhmRAO1QjhAAAAQLeE8JoI4WVCeCaEQzVCOAAAANAtIbwmQniZEJ4J4VCNEA4AAAB0SwiviRBeJoRnQjhUI4QDAAAA3RLCayKElwnhmRAO1QjhAAAAQLeE8JoI4WVCeCaEQzVCOAAAANAtIbwmQniZEJ4J4VCNEA4AAAB0SwiviRBeJoRnQjhUI4QDAAAA3RLCayKElwnhmRAO1QjhAAAAQLeE8JoI4WVCeCaEQzVCOAAAANAtIbwmQniZEJ4J4VCNEA4AAAB0SwiviRBeJoRnQjhUI4QDAAAA3RLCayKElwnhmRAO1QjhAAAAQLeE8JoI4WVCeCaEQzVCOAAAANAtIbwmQniZEJ4J4VCNEA4AAAB0SwiviRBeJoRnQjhUI4QDAAAA3RLCayKElwnhmRAO1QjhAAAAQLeE8JoI4WVCeCaEQzVCOAAAANAtIbwmQniZEJ4J4VCNEA4AAAB0SwiPiBdffDE2b94cV1xxRbTb7XjjjTc6x06ePBlbtmyJgYGBGB4ejj179nT1mkJ4mRCeCeFQjRAOAAAAdKvvQ/gLL7wQQ0NDMTk5Ga+//nq89tprsbi42Dk+Pj4eY2NjMTMzE0eOHImhoaGYmppa9nWF8DIhPBPCoRohHFhJG7/5TM/f93XNxm8+0+vlBQCAnuv7ED4yMhIPPfRQ8dji4mK02+04evRo57mJiYnYvn37sq8rhJcJ4ZkQDtUI4cBKEsIBAKDZ+jqEz83NRavVivHx8RgeHo41a9bE1q1bY25uLiIijh8/Hq1WKxYWFjq/ZnJyMjZu3Nh5vLS0VJwTb77d8x966pypl09c8Hu/2DQ5hP/Dv75SaU2aHsKrrIlpxlzsz8iVmEdf/HXP93ids7B4pue/h8b00zQ9hPd6fY0xptup+xrSGGMu5+HD6esQfuzYsWi1WjE1NRXz8/PxxhtvxHXXXRfj4+MRETE9PR2tVuu8jbZ3795Yt25d5/Hs7Gxxjr36es9/6KlzHn3+5xf83i82v3fjwz0/97rm648drrQmn/rqEz0/97pm9IFnK62Jaca8+eabtb7+959+ued7vM759Ru/6fnvoTH9NP/976d6/r6va/7730/1fH2NMabbqfsa0hhjLufhwxHCW604e/Y//vfzp59+Oq644oqI+I9PhJ97z/DJycnYsGHDsq/t1ihlTf5EuFuj5HFrFOrk1ijASmr6J8IBAKDf9XUIn5+fj1WrVsWxY8c6z01NTcXVV18dEb+9R/jq1avPu0f4zp07Y9u2bcu+thBeJoRnQjhUI4QDK0kIBwCAZuvrEB4R8aUvfSk2bdoUMzMzceLEiRgZGYm77767c3xsbCy2bt0as7OzMT09HcPDw7Fv375lX1cILxPCMyEcqhHCgZUkhAMAQLP1fQh/66234uabb47BwcG48sorY9euXfHuu+92js/MzMTo6Gi02+0YGhqK3bt3d/W6QniZEJ4J4VCNEA6sJCEcAACare9DeF2E8DIhPBPCoRohHFhJQjgAADSbEF4TIbxMCM+EcKhGCAdWkhAOAADNJoTXRAgvE8IzIRyqEcKBlSSEAwBAswnhNRHCy4TwTAiHaoRwYCUJ4QAA0GxCeE2E8DIhPBPCoRohHFhJQjgAADSbEF4TIbxMCM+EcKhGCAdWkhBONw7/6nTc+shLjZzb9v6s18sLAFArIbwmQniZEJ4J4VCNEA6sJCGcbnzvwC97/vtZ13zsC5O9Xl4AgFoJ4TURwsuE8EwIz+788dH4bxP/2sjZ9tDhSmtCJoQDK0kIpxtCOADA5UsIr4kQXiaEZ0J4dv0Dz/X83OuaP/3G/kprQiaEAytJCKcbQjgAwOVLCK+JEF4mhGdCeCaE0w0hPPu3E2/1/LzrnJ++eqqGnQS/JYTTDSEcAODyJYTXRAgvE8IzITwTwumGEJ4J4VCdEE43hHAAgMuXEF4TIbxMCM+E8EwIpxtCeCaEQ3VCON0QwgEALl9CeE2E8DIhPBPCMyGcbgjhmRAO1QnhdEMIBwC4fAnhNRHCy4TwTAjPhHC6IYRnQjhUJ4TTDSEcAODyJYTXRAgvE8IzITwTwumGEJ4J4WUf+8Jkz8+9rvnegV9WWhMyIZxuCOEAAJcvIbwmQniZEJ4J4ZkQTjeE8EwILxPC6YYQTjeEcACAy5cQXhMhvEwIz4TwTAinG0J4JoSXCeF0QwinG0I4AMDlSwiviRBeJoRnQngmhNMNITwTwsuEcLohhNMNIRwA4PIlhNdECC8TwjMhPBPC6YYQngnhZUI43RDC6YYQDgBw+RLCayKElwnhmRCeCeHZXY+/3PNzr2s+vv3RSmsihGdCeJkQTjeEcLohhAMr6U+/sb/n7/265voHnuv18gIkQnhNhPAyITwTwjMhPBPCMyE8E8LLhHC6IYTTDSEcWElCOMClJYTXRAgvE8IzITwTwjMhPBPCMyG8TAinG0I43RDCgZUkhANcWkJ4TYTwMiE8E8IzITwTwjMhPBPCy4Tw7N5/fSVufeSlRs6TR09UWhMhnG4I4cBKEsKzH7/0es+vJeqabz19fIV3EPBBCeE1EcLLhPBMCM+E8EwIz4TwTAgvE8KzP/7K4z0/97rmb374UqU1EcLphhAOrCQhPPvrH7zQ83Ova/7rHU+u8A4CPigh/P975513YtOmTbFp06bznj958mRs2bIlBgYGYnh4OPbs2dPV6wnhZUJ4JoRnQngmhGdCeCaElwnhmRCeCeF0QwgHVpIQngnhQJ2E8Ig4e/Zs3HjjjbFx48YUwsfHx2NsbCxmZmbiyJEjMTQ0FFNTU8u+phBeJoRnQngmhGdCeCaEZ0J4mRCeCeGZEE43hHBgJQnhmRAO1EkIj4hbbrklJiYm4uGHHz4vhC8uLka73Y6jR492npuYmIjt27cv+5pCeJkQngnhmRCeCeGZEJ4J4WVCeCaEZ0J49jc/fKnn517X/PFXHq+0JkI4sJKE8EwIB+rU9yF8165dcdNNN8XS0lIK4cePH49WqxULCwud5yYnJ2Pjxo3Lvq4QXiaEZ0J4JoRnQngmhGdCeJkQngnhmRCeCeGZEA6sJCE8E8KBOvV1CH/kkUdidHQ0zpw5ExGRQvj09HS0Wq1YWlrqPLd3795Yt25d5/G7775bnNdONjtGPDX9+gW/94tNk0P4vU8ei8XFxQ88g1/b1/Nzr2vG/vFQpTX5/H0Hen7udc2f7JqqtCY7H/tZz8+9rvn49kcr/Xnyzz/9956fe50zt/DOB16T6V+f6vl51zkHf/6bSnulySH8wZ/8vNKaNDmE7/i/hyutyYZ/+EnPz72u2fAPP6m0Jjv+7+Gen3td88dfebzSmjz4k5/3/Nzrmo99YbLSmhhTx1S5Xr4c5092TfX8vV/XfP6+A5XWZOwfD/X83Ouawa/t6/me+6hPr//suRyGD6evQ/hdd90Vq1evjna7He12O1atWhWtViva7XYsLCx0PhG+uLjY+TWTk5OxYcOGzuNTp04V55Vfnej5H7J1zmOHjl/we7/Y/N6ND/f83Ouab/z4pTh9+vQHnk999Ymen3tdc8ODByqtyf/a09wYcfVdT1Vak68+8mLPz72u+cMv7620JpPPvtLzc69z3pg5+YHX5NArzf6U/P6f/XulvdLkEL7nqelKa/JHt/yo5+de12z7waFKa7L+nuZ+Km/9Pfsrrcm2HzQ3RvzRLT+qtCZ7npru+bnXNR/7wmSlNTFmpWdubq7n53Cp5uq7nur5e7+u+V97flJpTW54sLkfivrUV5/o+Z77qE+VztRvw4fT1yH8/Ur3CF+9evV59wjfuXNnbNu2bdnXcmuUsiZ/ItytUfK4NUoet0bJ49Yo5XFrlDxujZLHrVHyuDVKHrdGyePWKHncGgUuPbdGydwaBaiTEH6O94fwiIixsbHYunVrzM7OxvT0dAwPD8e+ffuWfS0hvEwIz4TwTAjPhPBMCM+E8DIhPBPCMyE8E8IzIRxYSUJ4JoQDdRLCz1EK4TMzMzE6OhrtdjuGhoZi9+7dXb2WEF4mhGdCeCaEZ0J4JoRnQniZEJ4J4ZkQngnhmRAO1T3z89l46uUTjZxXZxcqrYkQngnhQJ2E8JoI4WVCeCaEZ0J4JoRnQngmhJcJ4ZkQngnhmRCeCeFQ3ce3P9rzfV7X3PX4y5XWRAjPhHCgTkJ4TYTwMiE8E8IzITwTwjMhPBPCy4TwTAjPhPBMCM+EcKhOCM+E8EwIB+okhNdECC8TwjMhPBPCMyE8E8IzIbxMCM+E8EwIz4TwTAiH6oTwTAjPhHCgTkJ4TYTwMiE8E8IzITwTwjMhPBPCy4TwTAjPhPBMCM+EcKhOCM+E8EwIB+okhNdECC8TwjMhPBPCMyE8E8IzIbxMCM+E8EwIz4TwTAiH6oTwTAjPhHCgTkJ4TYTwMiE8E8IzITwTwjMhPBPCy4TwTAjPhPBMCM+EcKhOCM+E8EwIB+okhNdECC8TwjMhPBPCMyE8E8IzIbxMCM+E8EwIz4TwTAiH6oTwTAjPhPDs+V+ejFsfeamRc8ePjlZaE6hKCK+JEF4mhGdCeCaEZ0J4JoRnQniZEJ4J4ZkQngnhmRAO1QnhmRCeCeHZN6d+3vNzr2t+f+yHldYEqhLCayKElwnhmRCeCeGZEJ4J4ZkQXiaEZ0J4JoRnQngmhEN1QngmhGdCeCaEw8oRwmsihJcJ4ZkQngnhmRCeCeGZEF4mhGdCeCaEZ0J4JoRDdUJ4JoRnQngmhMPKEcJrIoSXCeGZEJ4J4ZkQngnhmRBeJoRnQngmhGdCeCaEQ3VCeCaEZ0J4JoTDyhHCayKElwnhmRCeCeGZEJ4J4ZkQXiaEZ0J4JoRnQngmhEN1QngmhGdCeCaEw8oRwmsihJcJ4ZkQngnhmRCeCeGZEF4mhGdCeCaEZ0J4JoRDdUJ4JoRnQngmhMPKEcJrIoSXCeGZEJ4J4ZkQngnhmRBeJoRnQngmhGdCeCaEQ3VCeCaEZ0J4JoTDyhHCayKElwnhmRCeCeGZEJ4J4ZkQXiaEZ0J4JoRnQngmhEN1QngmhGdCeCaEw8oRwmsihJcJ4ZkQngnhmRCeCeGZEF4mhGdCeCaEZ0J4JoTTjbffPdvz388657HDr1VaFyE8E8IzITwTwmHlCOE1EcLLhPBMCM+E8EwIz4TwTAgvE8IzITwTwjMhPBPC6YYQXiaEZ0J4JoRnQjisHCG8JkJ4mRCeCeGZEJ4J4ZkQngnhZUJ4JoRnQngmhGdCON0QwsuE8EwIz4TwTAiHlSOE10QILxPCMyE8E8IzITwTwjMhvEwIz4TwTAjPhPBMCKcbQniZEJ4J4ZkQngnhsHKE8JoI4WVCeCaEZ0J4JoRnQngmhJcJ4ZkQngnhmRCeCeF0QwgvE8IzITwTwjMhHFaOEF4TIbxMCM+E8EwIz4TwTAjPhPAyITwTwjMhPBPCMyGcbgjhZUJ4JoRnQngmhMPKEcJrIoSXCeGZEJ4J4ZkQngnhmRBeJoRnQngmhGdCeCaE0w0hvEwIz4TwTAjPhHBYOUJ4TYTwMiE8E8IzITwTwjMhPBPCy4TwTAjPhPBMCM+EcLohhJcJ4ZkQngnhmRAOK0cIr4kQXiaEZ0J4JoRnQngmhGdCeJkQngnhmRCeCeGZEE43hPAyITwTwjMhPBPCYeUI4TURwsuE8EwIz4TwTAjPhPBMCC8TwjMhPBPCMyE8E8LphhBeJoRnQngmhGdCOKycvg/hd999d3zuc5+Ldrsda9euje985zvnHT958mRs2bIlBgYGYnh4OPbs2dPV6wrhZUJ4JoRnQngmhGdCeCaElwnhmRCeCeGZEJ4J4XRDCC8TwjMhPBPCMyEcVk7fh/CJiYk4ePBgzM7Oxv79+2NgYCCeffbZzvHx8fEYGxuLmZmZOHLkSAwNDcXU1NSyryuElwnhmRCeCeGZEJ4J4ZkQXiaEZ0J4JoRnQngmhNMNIbxMCM+E8EwIz4RwWDl9H8Lfb2RkJB544IGIiFhcXIx2ux1Hjx7tHJ+YmIjt27cv+zpCeJkQngnhmRCeCeGZEJ4J4WVCeCaEZ0J4JoRnQjjdEMLLhPBMCM+E8EwIh5UjhJ9jYWEhBgcH45lnfvvDwvHjx6PVasXCwkLnayYnJ2Pjxo3LvpYQXiaEZ0J4JoRnQngmhGdCeJkQngnhmRCeCeGZEJ69fvrtuPWRlxo7s28tfuA1EcLLhPBMCM+E8EwIh5UjhJ9jx44dsXnz5lhaWoqIiOnp6Wi1Wp3HERF79+6NdevWdR6/8847xfnVzJs9/wOlztn30q8v+L1fbJocwu/Z93KlNRn82r6en3tdc+P3nq+0Jp+/70DPz72u+ZNdU5XWZOdjP+v5udc1H9/+aKU1eeTQqz0/9zrn9FsLH3hNXvr32Z6fd51z4JUTlfZKk0P4A0+/UmlNVt36456fe11z00MvVlqT/7H76Z6fe13zP3Y/XWlNbnroxZ6fe12z6tYfV1qTB55+pefnXtd87AuTldbkwCsnen7udc5L/z77gdfk9FsLPT/vOueRQ69W2itNDuE7H/tZpTX5k11TPT/3uubz9x2otCY3fu/5np97XTP4tX2V1uSefc39UNTvj/2w0pr08/DhCOH/3x133BHr16+Pubm5znPvfSJ8cfE/PgUwOTkZGzZs6Dyem5srzvHXZnr+B0qd8+MXfnnB7/1i0+QQ/neP/6zSmnz6b/+l5+de12z5zsFKa7LpW839VN7Vdz1VaU3+9p8P9/zc65o//PLeSmvy0IHmfjLidzb/U/zm5OkPvCYvHH+j5+dd5zx99NeV9kqTQ/i3p16utCZ/dMuPen7udc2X/+mnldZk/T3N/VTe+nv2V1qTL//TT3t+7nXNH93yo0pr8u2p5saIj31hstKaPH301z0/9zrnheNvfOA1+c3J0z0/7zrnoQM/r7RX/vDLe3t+7nXN3/7z4UprcvVdT/X83OuaTd96ptKabPnOwZ6fe13z6b/9l0pr8nePN/dDUb8/9sNKa9LPw4fT9yH8zJkzsWPHjhgZGUkbanFxMVavXn3ePcJ37twZ27ZtW/Z13RqlrMkh3K1R8rg1Sh63Rsnj1ijlcWuUPG6NksetUfK4NUoet0bJ49YoeareGuWnr57q+bnXOf924q0PvCZujVLW5E+EuzVKHrdGyePWKHncGoVLra9D+JkzZ+L666+PG264IU6ePBnz8/MxPz9/3j3Bx8bGYuvWrTE7OxvT09MxPDwc+/btW/a1hfAyITwTwjMhPBPCMyE8E8LLhPBMCBPgKYAAABHPSURBVM+E8EwIz4TwTAjPhPAyITwTwjMhPBPCYeX0dQh/8803o9Vqpbnmmms6XzMzMxOjo6PRbrdjaGgodu/e3dVrC+FlQngmhGdCeCaEZ0J4JoSXCeGZEJ4J4ZkQngnhmRCeCeFlQngmhGdCeCaEw8rp6xBeJyG8TAjPhPBMCM+E8EwIz4TwMiE8E8IzITwTwjMhPBPCMyG8TAjPhPBMCM+EcFg5QnhNhPAyITwTwjMhPBPCMyE8E8LLhPBMCM+E8EwIz4TwTAjPhPAyITwTwjMhPBPCYeUI4TURwsuE8EwIz4TwTAjPhPBMCC8TwjMhPBPCMyE8E8IzITwTwsuE8EwIz4TwTAiHlSOE10QILxPCMyE8E8IzITwTwjMhvEwIz4TwTAjPhPBMCM+E8EwILxPCMyE8E8IzIRxWjhBeEyG8TAjPhPBMCM+E8EwIz4TwMiE8E8IzITwTwjMhPBPCMyG8TAjPhPBMCM+E8OyXs/Nx6yMvNXbmF89UWheWJ4TXRAgvE8IzITwTwjMhPBPCMyG8TAjPhPBMCM+E8EwIz4TwTAgvE8IzITwTwjMhPNt/7Dc9P/c65zdz71RaF5YnhNdECC8TwjMhPBPCMyE8E8IzIbxMCM+E8EwIz4TwTAjPhPBMCC8TwjMhPBPCMyE8E8KpSgiviRBeJoRnQngmhGdCeCaEZ0J4mRCeCeGZEJ4J4ZkQngnhmRBeJoRnQngmhGdCeCaEU5UQXhMhvEwIz4TwTAjPhPBMCM+E8DIhPBPCMyE8E8IzITwTwjMhvEwIz4TwTAjPhPBMCKcqIbwmQniZEJ4J4ZkQngnhmRCeCeFlQngmhGdCeCaEZ0J4JoRnQniZEJ4J4ZkQngnhmRBOVUJ4TYTwMiE8E8IzITwTwjMhPBPCy4TwTAjPhPBMCM+E8EwIz4TwMiE8E8IzITwTwjMhnKqE8JoI4WVCeCaEZ0J4JoRnQngmhJcJ4ZkQngnhmRCeCeGZEJ4J4WVCeCaEZ0J4JoRnQjhVCeE1EcLLhPBMCM+E8EwIz4TwTAgvE8IzITwTwjMhPBPCMyE8E8LLhPBMCM+E8EwIz4RwqhLCayKElwnhmRCeCeGZEJ4J4ZkQXiaEZ0J4JoRnQngmhGdCeCaElwnhmRCeCeGZEJ4J4VQlhNdECC8TwjMhPBPCMyE8E8IzIbxMCM+E8EwIz4TwTAjPhPBMCC8TwjMhPBPCMyE8E8KpSgiviRBeJoRnQngmhGdCeCaEZ0J4mRCeCeGZEJ4J4ZkQngnhmRBeJoRnQngmhGdCeCaEU5UQXhMhvEwIz4TwTAjPhPBMCM+E8DIhPBPCMyE8E8IzITwTwjMhvEwIz4TwTAjPhPBMCKcqIbwmQniZEJ4J4ZkQngnhmRCeCeFlQngmhGdCeCaEZ0J4JoRnQniZEJ4J4ZkQngnhmRBOVUJ4TYTwMiE8E8IzITwTwjMhPBPCy4TwTAjPhPBMCM+E8EwIz4TwMiE8E8IzITwTwjMhnKqE8JoI4WVCeCaEZ0J4JoRnQngmhJcJ4ZkQngnhmRCeCeGZEJ4J4WVCeCaEZ0J4JoRnQjhVCeE1EcLLhPBMCM+E8EwIz4TwTAgvE8IzITwTwjMhPBPCMyE8E8LLhPBMCM+E8EwIz4RwqhLCayKElwnhmRCeCeGZEJ4J4ZkQXiaEZ0J4JoRnQngmhGdCeCaElwnhmRCeCeGZEJ4J4VQlhNdECC8TwjMhPBPCMyE8E8IzIbxMCM+E8EwIz4TwTAjPhPBMCC8TwjMhPBPCMyE8E8KpSgiviRBeJoRnQngmhGdCeCaEZ0J4mRCeCeGZEJ4J4ZkQngnhmRBeJoRnQngmhGdCeCaEU5UQXhMhvEwIz4TwTAjPhPBMCM+E8DIhPBPCMyE8E8IzITwTwjMhvEwIz4TwTAjPhPBMCKcqIXwZJ0+ejC1btsTAwEAMDw/Hnj17uvp1QniZEJ4J4ZkQngnhmRCeCeFlQngmhGdCeCaEZ0J4JoRnQniZEJ4J4ZkQngnhmRBOVUL4MsbHx2NsbCxmZmbiyJEjMTQ0FFNTU8v+OiG8TAjPhPBMCM+E8EwIz4TwMiE8E8IzITwTwjMhPBPCMyG8TAjPhPBMCM+E8EwIpyoh/CIWFxej3W7H0aNHO89NTEzE9u3bl/21QniZEJ4J4ZkQngnhmRCeCeFlQngmhGdCeCaEZ0J4JoRnQniZEJ4J4ZkQngnhmRBOVUL4RRw/fjxarVYsLCx0npucnIyNGzcu+2uF8DIhPBPCMyE8E8IzITwTwsuE8EwIz4TwTAjPhPBMCM+E8DIhPBPCMyE8E8IzIZyqhPCLmJ6ejlarFUtLS53n9u7dG+vWres8vvfee4uz8+929/yNU+f89c7/c8Hv/WLzu3/1g56fe13zP//3tyutycfHvtvzc69rPrvjgUprMvjlB3p+7nVNa/y7ldZk/a339fzc65o/uOEfK63JDV/b0/Nzr3O+8fcffE2+suubPT/vOmf7xDcr7ZX/vLm5f/d8/qvfqrQm/+WL3+v5udc1f3rz/ZXW5BNferDn517XfOJLD1Zakz+9+f6en3td81+++L1Ka/L5r36r5+de1/znzT+otCbbJ5r9d89Xdn3wv3u+8ff39vy865wbvran0l75gxv+sefnXtesv/W+SmvSGm/uz4KDX672s+BndzT3Z8GPj1X7WfB//u9v9/zc65rf/atqf/f89c7/0/Nzr3N2/t3uC37vfDhC+EW894nwxcXFznOTk5OxYcOGzuN77rnHGGNMYR588MGen4MxxhhjjLm8xjWkMcZcePhwhPCLWFxcjNWrV593j/CdO3fGtm3benhWAAAAAAB8EEL4MsbGxmLr1q0xOzsb09PTMTw8HPv27ev1aQEAAAAA0CUhfBkzMzMxOjoa7XY7hoaGYvfu3b0+JQAAAAAAPgAh/CPu3XffjVtvvTVuv/324vEXX3wxNm/eHFdccUW02+144403Oseuvfba+PSnPx3z8/Od586cORNXXXVVfOpTn6r93Ll0LrZPfvGLX8TmzZvjyiuvjM985jNx++23n3ffe/ukP9x9993xuc99Ltrtdqxduza+853vnHf85MmTsWXLlhgYGIjh4eHYs2fPecftk/6x3F5Z7ri90h+W2wfveeedd2LTpk2xadOm8563T/pDN/vEtSzL7RPXsrznvvvuiz//8z+PgYGBGBwcjJtuuineeuutznHXs0Qsv09cyxKx/D55j2vZ5hHCP8KefPLJuOqqq+ITn/hEMXC+8MILMTQ0FJOTk/H666/Ha6+9li4Kr7766rj//vs7zz366KOxdu1ab8wGWW6fXHvttZ09MDc3FyMjI+f9S8P2SX+YmJiIgwcPxuzsbOzfvz8GBgbi2Wef7RwfHx+PsbGxmJmZiSNHjsTQ0FBMTU11jtsn/WO5vbLccXulPyy3DyIizp49GzfeeGNs3Lix+MODfdJ8y+0T17JEdPf3imtZIiIefvjhOHz4cJw6dSpeeeWV+LM/+7P49re/3TnuepaI5feJa1kilt8nEa5lm0oIvwzceeedxcA5MjISDz300AV/3bXXXhvf/e5347Of/Wy8++67ERGxYcOG+O53v+uN2UClfXL27NlYtWpVHDx48Lyv27FjR+exfdKfRkZG4oEHHoiI3/7DwO12+7x/GHhiYiK2b9/eeWyf9K9z90o3x+2V/lTaJ7fccktMTEzEww8/XPzhwT7pP+/fJ65lKTl3n7iWpeTs2bNx7NixuOaaa+LAgQMR4XqWrLRPSlzL9reL7RPXss0khF8GSoFzbm4uWq1WjI+Px/DwcKxZsya2bt0ac3Nzna+59tpr4+DBg/GXf/mX8fDDD8dzzz0X1113Xbz44ovemA10of9gMjExEUNDQ/H9738/jh49GuvWrTvvAtE+6T8LCwsxODgYzzzzTEREHD9+PFqtViwsLHS+ZnJyMjZu3Nh5bJ/0p/fvlW6O2yv9p7QPdu3aFTfddFMsLS1d8IcH+6S/vH+fuJalpPTniWtZ3q/VasXq1avP+w9prmd5v9I+eT/Xslxon7iWbS4h/DJQCpzHjh2LVqsVU1NTMT8/H2+88UZcd911MT4+3vma996Yzz77bKxbty5GR0dj//793pgNdaEQ/vzzz8f69etj+/btsWbNmhgbGzvvPlb2Sf/ZsWNHbN68OZaWliIiYnp6OlqtVudxRMTevXtj3bp1ncf2SX96/17p5ri90n/evw8eeeSRGB0djTNnzkREXPSHB/ukf7x/n7iWpaT094prWd7vvU9wXn311fHEE09EhOtZstI+eT/XspT2iWvZZhPCLwMXC+Fnz57tPPf000/HFVdc0Xn83hsz4rf/i8Zf/MVfRER4YzZUaZ+cOnUqPvnJT8arr74aERGnT5+O66+/PsbGxjpfY5/0lzvuuCPWr19/3ifu3vsEzbn3ZZ2cnIwNGzZ0Htsn/ae0V7o5bq/0l9I+uOuuu2L16tXRbrej3W7HqlWrotVqRbvd7nxSzz7pL6V94lqW9yvtE9eyXMzXv/71zn88cz3LhZy7T87lWpZznbtPXMs2mxB+GSgFzvn5+Vi1alUcO3as89zU1FRcffXVncfnvjEPHToUzz33XER4YzZVaZ8cOnQo1qxZc95ze/fujbVr13Ye2yf94cyZM7Fjx44YGRlJF3uLi4uxevXq8/434507d8a2bds6j+2T/nGxvdLNcXulPyy3D851sU/RRNgnTXaxfeJalvdcbJ+4luVibrvttrjpppsiwvUsF3buPolwLUvZ+/fJuVzLNosQfhm40C0vvvSlL8WmTZtiZmYmTpw4ESMjI3H33Xd3jp/7xjyXN2YzXehe8oODg3H//ffH0tJSzM/Px+joaNx2222dr7FPmu/MmTNx/fXXxw033BAnT56M+fn5mJ+fP+8eimNjY7F169aYnZ2N6enpGB4ejn379nWO2yf9Ybm90s1eslear5t9cK7lfng4l33SHN3sE9eyLLdPXMvynrNnz8YXv/jFOHDgQMzMzMQTTzwRa9asif3793e+xvUsy+0T17JEdPfnyblcyzaLEP4R9thjj8Xg4GAMDAxEu92OwcHBePzxxzvH33rrrbj55ptjcHAwrrzyyti1a1fnX6uN8MbsF8vtk8OHD8fIyEgMDw/HVVddFXfccUe8/fbbneP2SfO9+eab0Wq10lxzzTWdr5mZmYnR0dFot9sxNDQUu3fvPu817JP+sNxe6WYv2SvN180+OJcfHvpTN/vEtSzd7BPXskRELC0txc033xxr166Ndrsd69ati0ceeeS8r3E9y3L7xLUsEd39eXIu17LNIoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBoQjgAAAAAAI0mhAMAAAAA0GhCOAAAAAAAjSaEAwAAAADQaEI4AAAAAACNJoQDAAAAANBo/w/KYPL2LrQhqwAAAABJRU5ErkJggg==",
      "text/html": [
       "<div id=\"764c47a5-8878-4be7-bda4-fcdc1fa2cde5\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"764c47a5-8878-4be7-bda4-fcdc1fa2cde5\", [{\"x\": [16000000, 17000000, 18000000, 19000000, 20000000, 21000000, 22000000, 23000000, 24000000, 25000000, 26000000, 27000000, 28000000, 29000000, 30000000, 31000000, 32000000, 33000000, 34000000], \"y\": [101, 100, 85, 53, 46, 85, 52, 49, 44, 62, 42, 54, 21, 29, 38, 46, 44, 34, 15], \"type\": \"bar\", \"uid\": \"4b5f3375-4256-4850-bbc6-c66493cdf1b5\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\"})});</script><script type=\"text/javascript\">window.addEventListener(\"resize\", function(){window._Plotly.Plots.resize(document.getElementById(\"764c47a5-8878-4be7-bda4-fcdc1fa2cde5\"));});</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<div id=\"764c47a5-8878-4be7-bda4-fcdc1fa2cde5\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"764c47a5-8878-4be7-bda4-fcdc1fa2cde5\", [{\"x\": [16000000, 17000000, 18000000, 19000000, 20000000, 21000000, 22000000, 23000000, 24000000, 25000000, 26000000, 27000000, 28000000, 29000000, 30000000, 31000000, 32000000, 33000000, 34000000], \"y\": [101, 100, 85, 53, 46, 85, 52, 49, 44, 62, 42, 54, 21, 29, 38, 46, 44, 34, 15], \"type\": \"bar\", \"uid\": \"4b5f3375-4256-4850-bbc6-c66493cdf1b5\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\"})});</script><script type=\"text/javascript\">window.addEventListener(\"resize\", function(){window._Plotly.Plots.resize(document.getElementById(\"764c47a5-8878-4be7-bda4-fcdc1fa2cde5\"));});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly.offline as py\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import get_plotlyjs, init_notebook_mode\n",
    "\n",
    "pandasHistogramDF = oskar.histogram(df,\"start\",1000000).toPandas()\n",
    "\n",
    "init_notebook_mode(connected=True)\n",
    "\n",
    "data = [go.Bar(x=pandasHistogramDF[\"start\"], y=pandasHistogramDF[\"count\"])]\n",
    "py.iplot(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** Note that the region covers from the self x-axis value to the next one*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** If you are using jupyter lab you will need to download this extension https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/plotly-extension to visualize the plot*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Identity by state\n",
    "Some other interesting stats we may want to acces is IBS coefficient, wich could give us important information about variant similarity. For that we count on [<span style=\"color:#FF1493\"> **ibs** </span>] transformer:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+------------------+------------+--------+----+\n",
      "|        samplePair|          distance|      counts|variants|skip|\n",
      "+------------------+------------------+------------+--------+----+\n",
      "|[NA12877, NA12878]|0.8928571428571429|[0, 57, 209]|     266| 734|\n",
      "|[NA12877, NA12879]|0.9155629139072847|[0, 51, 251]|     302| 698|\n",
      "|[NA12877, NA12880]|0.9028662420382165|[0, 61, 253]|     314| 686|\n",
      "|[NA12877, NA12881]|0.9116719242902208|[0, 56, 261]|     317| 683|\n",
      "|[NA12877, NA12882]|0.8934707903780069|[0, 62, 229]|     291| 709|\n",
      "|[NA12877, NA12883]| 0.891156462585034|[0, 64, 230]|     294| 706|\n",
      "|[NA12877, NA12884]|0.8904109589041096|[0, 64, 228]|     292| 708|\n",
      "|[NA12877, NA12885]|0.9115755627009646|[0, 55, 256]|     311| 689|\n",
      "|[NA12877, NA12886]| 0.909967845659164|[0, 56, 255]|     311| 689|\n",
      "|[NA12877, NA12887]|0.9096989966555183|[0, 54, 245]|     299| 701|\n",
      "+------------------+------------------+------------+--------+----+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.ibs(df).show(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In case we don't specify any samples it will run the whole dataframe samples, but we may want to check similarity between concrete individuals:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+------------------+------------+--------+----+\n",
      "|        samplePair|          distance|      counts|variants|skip|\n",
      "+------------------+------------------+------------+--------+----+\n",
      "|[NA12877, NA12878]|0.8928571428571429|[0, 57, 209]|     266| 734|\n",
      "|[NA12877, NA12879]|0.9155629139072847|[0, 51, 251]|     302| 698|\n",
      "|[NA12878, NA12879]|0.8857615894039735|[0, 69, 233]|     302| 698|\n",
      "+------------------+------------------+------------+--------+----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.ibs(df, samples=[\"NA12877\", \"NA12878\", \"NA12879\"]).show(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<span style=\"color:yellow\"> **COMENTO LOS POSIBLES PARÁMETROS??** </span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "root\n",
      " |-- samplePair: array (nullable = true)\n",
      " |    |-- element: string (containsNull = true)\n",
      " |-- distance: double (nullable = true)\n",
      " |-- counts: array (nullable = true)\n",
      " |    |-- element: integer (containsNull = false)\n",
      " |-- variants: integer (nullable = true)\n",
      " |-- skip: integer (nullable = true)\n",
      "\n",
      "['NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12877', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12878', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12879', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12880', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12881', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12882', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12883', 'NA12884', 'NA12884', 'NA12884', 'NA12884', 'NA12884', 'NA12884', 'NA12884', 'NA12884', 'NA12884', 'NA12885', 'NA12885', 'NA12885', 'NA12885', 'NA12885', 'NA12885', 'NA12885', 'NA12885', 'NA12886', 'NA12886', 'NA12886', 'NA12886', 'NA12886', 'NA12886', 'NA12886', 'NA12887', 'NA12887', 'NA12887', 'NA12887', 'NA12887', 'NA12887', 'NA12888', 'NA12888', 'NA12888', 'NA12888', 'NA12888', 'NA12889', 'NA12889', 'NA12889', 'NA12889', 'NA12890', 'NA12890', 'NA12890', 'NA12891', 'NA12891', 'NA12892']\n"
     ]
    }
   ],
   "source": [
    "oskar.ibs(df).printSchema()\n",
    "OskarIBS = oskar.ibs(df).toPandas()\n",
    "samples = [sample[0] for sample in OskarIBS.samplePair]\n",
    "print(samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Advanced variant filtering**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Hardy Weinberg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------------+\n",
      "|             id|                 HWE|\n",
      "+---------------+--------------------+\n",
      "|22:16054454:C:T|                 1.0|\n",
      "|22:16065809:T:C|                 1.0|\n",
      "|22:16077310:T:A|  0.9254727474972191|\n",
      "|22:16080499:A:G|                 1.0|\n",
      "|22:16084621:T:C|                 1.0|\n",
      "|22:16091610:G:T|                 1.0|\n",
      "|22:16096040:G:A|  0.4746014089729329|\n",
      "|22:16099957:C:T|0.016007636455477054|\n",
      "|22:16100462:A:G|0.001011008618240...|\n",
      "|22:16105660:G:A|  0.3037449017426771|\n",
      "|22:16112391:G:A| 0.17718452601656157|\n",
      "|22:16114913:A:T|  0.6855353685164587|\n",
      "|22:16127471:A:-|                 1.0|\n",
      "|22:16134019:G:T| 0.17718452601656157|\n",
      "|22:16138943:C:G|  0.6855353685164587|\n",
      "|22:16144239:T:C| 0.18181818181818182|\n",
      "|22:16147398:G:A|0.001011224592982...|\n",
      "|22:16149692:G:T| 0.08884238232789762|\n",
      "|22:16195955:G:A|0.007751066073178...|\n",
      "|22:16196041:C:T|  0.6855353685164587|\n",
      "|22:16202382:C:T|0.001011008618240...|\n",
      "|22:16212765:C:T|                 1.0|\n",
      "|22:16223351:C:T|                 1.0|\n",
      "|22:16229483:C:T|                 1.0|\n",
      "|22:16232104:G:A|                 1.0|\n",
      "|22:16240905:C:T|                 1.0|\n",
      "|22:16252532:A:G| 0.06060606060606061|\n",
      "|22:16289841:A:-|  0.3547151155604993|\n",
      "|22:16302286:G:T|                 1.0|\n",
      "|22:16308379:G:A|                 1.0|\n",
      "+---------------+--------------------+\n",
      "only showing top 30 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.hardyWeinberg(df,\"hgvauser@platinum:illumina_platinum\").select(\"id\", \"HWE\").show(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inbreeding coefficient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+-------------------+-----------+------------------+--------------+\n",
      "|SampleId|                  F|ObservedHom|       ExpectedHom|GenotypesCount|\n",
      "+--------+-------------------+-----------+------------------+--------------+\n",
      "| NA12877|-1.0857581722788996|         70|233.97577702999115|           385|\n",
      "| NA12878|-1.1024114888695444|         69|244.65916746854782|           404|\n",
      "| NA12879|-1.1890914293957586|         69| 247.7093403339386|           398|\n",
      "| NA12880|-1.1013660394101679|         71|248.15224742889404|           409|\n",
      "| NA12881|-1.1560267972581504|         65| 252.6643579006195|           415|\n",
      "| NA12882|-1.0112382612189488|         76| 224.8269881606102|           372|\n",
      "| NA12883|-1.0602574055431329|         67|229.62110525369644|           383|\n",
      "| NA12884|-1.0340014363992485|         74|224.47404664754868|           370|\n",
      "| NA12885|-1.1105665251221366|         78| 254.8010356426239|           414|\n",
      "| NA12886| -1.067867784696387|         72|244.48096668720245|           406|\n",
      "+--------+-------------------+-----------+------------------+--------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df2 = oskar.stats(df, studyId=\"hgvauser@platinum:illumina_platinum\", missingAsReference=True)\n",
    "oskar.inbreedingCoefficient(df2).show(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mendelian error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+------------------+--------------+\n",
      "|                id|mendelianError|\n",
      "+------------------+--------------+\n",
      "|22:19748211:CCCC:-|             1|\n",
      "+------------------+--------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.mendel(df, \"NA12877\", \"NA12878\", \"NA12879\").select(\"id\", \"mendelianError\").filter(col(\"mendelianError\") != \"0\").show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
