{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Statistics tutorial**\n",
    "\n",
    "In this Jupyter tutorial we will show some examples of how to extract statistics from our data. These methods are implemented as part of the functionality of PyOskar API using the _transformers_ Spark model. These kind of functions depend directly from an _Oskar()_ object, and they will return a whole dataframe with the due transformation applied.\n",
    "> <span style=\"color:#ff6600\"> **PyOskar is integrated in PySpark default API. It is possible to manage data simultaneously with functions from both libraries in case we want to select specific fields, visualize the dataframe, filter the output or performing other operations besides the transformation.** </span>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we need to import the PyOskar and PySpark modules. Second, we need to create an instance of the _Oskar()_ object, from which depends a big part of the functionality. Finally, we must use the _load()_ transformer pointing to where the parquet file is stored to convert our data into a DataFrame _df_, and we are ready to start playing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyoskar.core import Oskar\n",
    "from pyoskar.sql import *\n",
    "from pyoskar.analysis import *\n",
    "from pyspark.sql.functions import *\n",
    "\n",
    "oskar = Oskar(spark)\n",
    "df = oskar.load(\"./data/platinum_chr22.small.parquet\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can use PySpark _show()_ method to print the data from _df_. This is how our testing dataframe looks like. As you can see for this tutorial we have selected a small dataset from Illumina Platinum Genomes with 1,000 random variants from chromosome 22."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Print first 20 variants:\n",
      "+---------------+-----+----------+--------+--------+---------+---------+------+----+------+-----+----+--------------------+--------------------+\n",
      "|             id|names|chromosome|   start|     end|reference|alternate|strand|  sv|length| type|hgvs|             studies|          annotation|\n",
      "+---------------+-----+----------+--------+--------+---------+---------+------+----+------+-----+----+--------------------+--------------------+\n",
      "|22:16054454:C:T|   []|        22|16054454|16054454|        C|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16054454, 16...|\n",
      "|22:16065809:T:C|   []|        22|16065809|16065809|        T|        C|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16065809, 16...|\n",
      "|22:16077310:T:A|   []|        22|16077310|16077310|        T|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16077310, 16...|\n",
      "|22:16080499:A:G|   []|        22|16080499|16080499|        A|        G|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16080499, 16...|\n",
      "|22:16084621:T:C|   []|        22|16084621|16084621|        T|        C|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16084621, 16...|\n",
      "|22:16091610:G:T|   []|        22|16091610|16091610|        G|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16091610, 16...|\n",
      "|22:16096040:G:A|   []|        22|16096040|16096040|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16096040, 16...|\n",
      "|22:16099957:C:T|   []|        22|16099957|16099957|        C|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16099957, 16...|\n",
      "|22:16100462:A:G|   []|        22|16100462|16100462|        A|        G|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16100462, 16...|\n",
      "|22:16105660:G:A|   []|        22|16105660|16105660|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16105660, 16...|\n",
      "|22:16112391:G:A|   []|        22|16112391|16112391|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16112391, 16...|\n",
      "|22:16114913:A:T|   []|        22|16114913|16114913|        A|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16114913, 16...|\n",
      "|22:16127471:A:-|   []|        22|16127471|16127471|        A|         |     +|null|     1|INDEL|  []|[[hgvauser@platin...|[22, 16127471, 16...|\n",
      "|22:16134019:G:T|   []|        22|16134019|16134019|        G|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16134019, 16...|\n",
      "|22:16138943:C:G|   []|        22|16138943|16138943|        C|        G|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16138943, 16...|\n",
      "|22:16144239:T:C|   []|        22|16144239|16144239|        T|        C|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16144239, 16...|\n",
      "|22:16147398:G:A|   []|        22|16147398|16147398|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16147398, 16...|\n",
      "|22:16149692:G:T|   []|        22|16149692|16149692|        G|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16149692, 16...|\n",
      "|22:16195955:G:A|   []|        22|16195955|16195955|        G|        A|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16195955, 16...|\n",
      "|22:16196041:C:T|   []|        22|16196041|16196041|        C|        T|     +|null|     1|  SNV|  []|[[hgvauser@platin...|[22, 16196041, 16...|\n",
      "+---------------+-----+----------+--------+--------+---------+---------+------+----+------+-----+----+--------------------+--------------------+\n",
      "only showing top 20 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(\"Print first 20 variants:\")\n",
    "df.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOTE: With _df.printSchema()_ command you can check the dataset hierarchy and all its fields."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of variants:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1000"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(\"Total number of variants:\")\n",
    "df.count()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats per variant\n",
    "Here we present the genotypes of the variants from which the later stats will come out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "|id             |genotypes                                                                                                                |\n",
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "|22:16054454:C:T|[[[./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [./.], [./.], [./.]]]|\n",
      "|22:16065809:T:C|[[[0/1], [./.], [./.], [./.], [0/1], [./.], [0/1], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [./.], [./.], [0/1]]]|\n",
      "|22:16077310:T:A|[[[./.], [0/1], [0/1], [./.], [./.], [0/1], [./.], [0/1], [./.], [0/1], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16080499:A:G|[[[0/1], [./.], [0/1], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [./.], [./.], [./.]]]|\n",
      "|22:16084621:T:C|[[[./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16091610:G:T|[[[./.], [./.], [./.], [./.], [./.], [./.], [./.], [0/1], [./.], [./.], [./.], [0/1], [./.], [./.], [0/1], [./.], [./.]]]|\n",
      "|22:16096040:G:A|[[[./.], [./.], [./.], [./.], [./.], [./.], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [0/1]]]|\n",
      "|22:16099957:C:T|[[[0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [1/1], [0/1], [1/1], [1/1], [0/1], [0/1]]]|\n",
      "|22:16100462:A:G|[[[0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1], [./.], [0/1], [0/1], [0/1], [0/1], [0/1], [0/1]]]|\n",
      "|22:16105660:G:A|[[[0/1], [./.], [0/1], [./.], [0/1], [./.], [0/1], [0/1], [./.], [0/1], [./.], [0/1], [0/1], [./.], [0/1], [./.], [0/1]]]|\n",
      "+---------------+-------------------------------------------------------------------------------------------------------------------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.select(df.id, df.studies.samplesData.alias(\"genotypes\")).show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this case we are applying a transformation that adds new data inside the _stats_ field, which is empty before we do anything:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16065809:T:C|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16077310:T:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16080499:A:G|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16084621:T:C|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16091610:G:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16096040:G:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16099957:C:T|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16100462:A:G|null       |null          |null          |null              |null                |null         |null         |\n",
      "|22:16105660:G:A|null       |null          |null          |null              |null                |null         |null         |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "df.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\")\\\n",
    "    .show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we proceed to call the _stats()_ transformer through our _oskar_ instance. That means that our dataframe will go through some kind of operation. It is important to know how transformers work in order to correctly access the new information.\n",
    "<br>\n",
    "Usage:\n",
    "```python\n",
    "stats(df[DataFrame], studyId[str]=None, cohort[str]=None, samples[str]=None, missingAsReference[bool]=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|2          |1             |1             |32                |16                  |0.5          |0.5          |\n",
      "|22:16065809:T:C|12         |6             |6             |22                |11                  |0.5          |0.5          |\n",
      "|22:16077310:T:A|14         |7             |7             |20                |10                  |0.5          |0.5          |\n",
      "|22:16080499:A:G|8          |4             |4             |26                |13                  |0.5          |0.5          |\n",
      "|22:16084621:T:C|10         |5             |5             |24                |12                  |0.5          |0.5          |\n",
      "|22:16091610:G:T|6          |3             |3             |28                |14                  |0.5          |0.5          |\n",
      "|22:16096040:G:A|18         |9             |9             |16                |8                   |0.5          |0.5          |\n",
      "|22:16099957:C:T|34         |14            |20            |0                 |0                   |0.4117647    |0.5882353    |\n",
      "|22:16100462:A:G|32         |16            |16            |2                 |1                   |0.5          |0.5          |\n",
      "|22:16105660:G:A|20         |10            |10            |14                |7                   |0.5          |0.5          |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "statsDF = oskar.stats(df)\n",
    "statsDF.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\")\\\n",
    "    .show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since our data has some empty genotypes, we would like to treat those missing values as if they were the reference. For that issue we dispose of _missingAsReference_ parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|id             |alleleCount|refAlleleCount|altAlleleCount|missingAlleleCount|missingGenotypeCount|refAlleleFreq|altAlleleFreq|\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "|22:16054454:C:T|34         |33            |1             |0                 |0                   |0.9705882    |0.029411765  |\n",
      "|22:16065809:T:C|34         |28            |6             |0                 |0                   |0.8235294    |0.1764706    |\n",
      "|22:16077310:T:A|34         |27            |7             |0                 |0                   |0.7941176    |0.20588236   |\n",
      "|22:16080499:A:G|34         |30            |4             |0                 |0                   |0.88235295   |0.11764706   |\n",
      "|22:16084621:T:C|34         |29            |5             |0                 |0                   |0.85294116   |0.14705883   |\n",
      "|22:16091610:G:T|34         |31            |3             |0                 |0                   |0.9117647    |0.0882353    |\n",
      "|22:16096040:G:A|34         |25            |9             |0                 |0                   |0.7352941    |0.2647059    |\n",
      "|22:16099957:C:T|34         |14            |20            |0                 |0                   |0.4117647    |0.5882353    |\n",
      "|22:16100462:A:G|34         |18            |16            |0                 |0                   |0.5294118    |0.47058824   |\n",
      "|22:16105660:G:A|34         |24            |10            |0                 |0                   |0.7058824    |0.29411766   |\n",
      "+---------------+-----------+--------------+--------------+------------------+--------------------+-------------+-------------+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "missingAsReferenceStatsDF = oskar.stats(df, missingAsReference = True)\n",
    "missingAsReferenceStatsDF.selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"alleleCount\", \"refAlleleCount\", \"altAlleleCount\", \"missingAlleleCount\", \"missingGenotypeCount\", \"refAlleleFreq\", \"altAlleleFreq\")\\\n",
    "    .show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Other statistics that our method provides are the next ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "|id             |genotypeCount        |genotypeFreq                                     |mafAllele|maf        |mgfGenotype|mgf|\n",
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "|22:16054454:C:T|[0/0 -> 16, 0/1 -> 1]|[0/0 -> 0.9411765, 0/1 -> 0.05882353, 1/1 -> 0.0]|T        |0.029411765|1/1        |0.0|\n",
      "|22:16065809:T:C|[0/0 -> 11, 0/1 -> 6]|[0/0 -> 0.64705884, 0/1 -> 0.3529412, 1/1 -> 0.0]|C        |0.1764706  |1/1        |0.0|\n",
      "|22:16077310:T:A|[0/0 -> 10, 0/1 -> 7]|[0/0 -> 0.5882353, 0/1 -> 0.4117647, 1/1 -> 0.0] |A        |0.20588236 |1/1        |0.0|\n",
      "|22:16080499:A:G|[0/0 -> 13, 0/1 -> 4]|[0/0 -> 0.7647059, 0/1 -> 0.23529412, 1/1 -> 0.0]|G        |0.11764706 |1/1        |0.0|\n",
      "|22:16084621:T:C|[0/0 -> 12, 0/1 -> 5]|[0/0 -> 0.7058824, 0/1 -> 0.29411766, 1/1 -> 0.0]|C        |0.14705883 |1/1        |0.0|\n",
      "|22:16091610:G:T|[0/0 -> 14, 0/1 -> 3]|[0/0 -> 0.8235294, 0/1 -> 0.1764706, 1/1 -> 0.0] |T        |0.0882353  |1/1        |0.0|\n",
      "|22:16096040:G:A|[0/0 -> 8, 0/1 -> 9] |[0/0 -> 0.47058824, 0/1 -> 0.5294118, 1/1 -> 0.0]|A        |0.2647059  |1/1        |0.0|\n",
      "|22:16099957:C:T|[0/1 -> 14, 1/1 -> 3]|[0/0 -> 0.0, 0/1 -> 0.8235294, 1/1 -> 0.1764706] |C        |0.4117647  |0/0        |0.0|\n",
      "|22:16100462:A:G|[0/0 -> 1, 0/1 -> 16]|[0/0 -> 0.05882353, 0/1 -> 0.9411765, 1/1 -> 0.0]|G        |0.47058824 |1/1        |0.0|\n",
      "|22:16105660:G:A|[0/0 -> 7, 0/1 -> 10]|[0/0 -> 0.4117647, 0/1 -> 0.5882353, 1/1 -> 0.0] |A        |0.29411766 |1/1        |0.0|\n",
      "+---------------+---------------------+-------------------------------------------------+---------+-----------+-----------+---+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.stats(df, missingAsReference = True).selectExpr(\"id\", \"studies[0].stats.ALL as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"genotypeCount\", \"genotypeFreq\", \"mafAllele\", \"maf\", \"mgfGenotype\", \"mgf\").show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are other options we could execute:\n",
    " - Getting stats from specific individuals. In that case we need to provide a list of samples.\n",
    " - Specify the name we want to give to the stats cohort, which by default is \"ALL\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+--------------------+------------------------------------+---------+----+-----------+---+\n",
      "|id             |genotypeCount       |genotypeFreq                        |mafAllele|maf |mgfGenotype|mgf|\n",
      "+---------------+--------------------+------------------------------------+---------+----+-----------+---+\n",
      "|22:16054454:C:T|[0/0 -> 10]         |[0/0 -> 1.0, 0/1 -> 0.0, 1/1 -> 0.0]|T        |0.0 |0/1        |0.0|\n",
      "|22:16065809:T:C|[0/0 -> 7, 0/1 -> 3]|[0/0 -> 0.7, 0/1 -> 0.3, 1/1 -> 0.0]|C        |0.15|1/1        |0.0|\n",
      "|22:16077310:T:A|[0/0 -> 5, 0/1 -> 5]|[0/0 -> 0.5, 0/1 -> 0.5, 1/1 -> 0.0]|A        |0.25|1/1        |0.0|\n",
      "|22:16080499:A:G|[0/0 -> 8, 0/1 -> 2]|[0/0 -> 0.8, 0/1 -> 0.2, 1/1 -> 0.0]|G        |0.1 |1/1        |0.0|\n",
      "|22:16084621:T:C|[0/0 -> 7, 0/1 -> 3]|[0/0 -> 0.7, 0/1 -> 0.3, 1/1 -> 0.0]|C        |0.15|1/1        |0.0|\n",
      "|22:16091610:G:T|[0/0 -> 9, 0/1 -> 1]|[0/0 -> 0.9, 0/1 -> 0.1, 1/1 -> 0.0]|T        |0.05|1/1        |0.0|\n",
      "|22:16096040:G:A|[0/0 -> 7, 0/1 -> 3]|[0/0 -> 0.7, 0/1 -> 0.3, 1/1 -> 0.0]|A        |0.15|1/1        |0.0|\n",
      "|22:16099957:C:T|[0/1 -> 10]         |[0/0 -> 0.0, 0/1 -> 1.0, 1/1 -> 0.0]|C        |0.5 |0/0        |0.0|\n",
      "|22:16100462:A:G|[0/1 -> 10]         |[0/0 -> 0.0, 0/1 -> 1.0, 1/1 -> 0.0]|A        |0.5 |0/0        |0.0|\n",
      "|22:16105660:G:A|[0/0 -> 4, 0/1 -> 6]|[0/0 -> 0.4, 0/1 -> 0.6, 1/1 -> 0.0]|A        |0.3 |1/1        |0.0|\n",
      "+---------------+--------------------+------------------------------------+---------+----+-----------+---+\n",
      "only showing top 10 rows\n",
      "\n"
     ]
    }
   ],
   "source": [
    "samples = oskar.metadata.samples(df)[\"hgvauser@platinum:illumina_platinum\"]\n",
    "tenSamples = list(samples[0:10])\n",
    "oskar.stats(df, studyId = \"hgvauser@platinum:illumina_platinum\", cohort = \"Caucasians\", samples = tenSamples, missingAsReference = True)\\\n",
    "    .selectExpr(\"id\", \"studies[0].stats.Caucasians as stats\").select(\"id\", \"stats.*\")\\\n",
    "    .select(\"id\", \"genotypeCount\", \"genotypeFreq\", \"mafAllele\", \"maf\", \"mgfGenotype\", \"mgf\").show(10, truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Global Stats\n",
    "We could also want to use _globalStats()_ that provides statistics attached to the whole dataframe.\n",
    "<br>\n",
    "Usage:\n",
    "```python\n",
    "globalStats(df[DataFrame], studyId[str]=None, fileId[str]=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some fields will be given as *int* and *float*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "|numVariants|numSamples|numPass|tiTvRatio|meanQuality|stdDevQuality|\n",
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "|       1000|         0|   4016|1.3342036|    773.252|    765.54596|\n",
      "+-----------+----------+-------+---------+-----------+-------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe = oskar.globalStats(df)\n",
    "globalStatsTransformationDataframe.select(\"numVariants\", \"numSamples\", \"numPass\", \"tiTvRatio\", \"meanQuality\", \"stdDevQuality\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Others are returned as *array*, although in this case it is empty:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+---------------+\n",
      "|numRareVariants|\n",
      "+---------------+\n",
      "|             []|\n",
      "+---------------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe.select(\"numRareVariants\").show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And others will be given as _map_. In order to show it here we previously converted it into a new DataFrame by using _explode_ PySpark command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+-----------------+-----+\n",
      "|variantTypeCounts|value|\n",
      "+-----------------+-----+\n",
      "|              SNV|  894|\n",
      "|            INDEL|  106|\n",
      "+-----------------+-----+\n",
      "\n",
      "+----------------------------------+-----+\n",
      "|variantBiotypeCounts              |value|\n",
      "+----------------------------------+-----+\n",
      "|IG_C_pseudogene                   |1    |\n",
      "|IG_V_pseudogene                   |17   |\n",
      "|protein_coding                    |500  |\n",
      "|rRNA                              |4    |\n",
      "|retained_intron                   |192  |\n",
      "|nonsense_mediated_decay           |166  |\n",
      "|IG_V_gene                         |18   |\n",
      "|snRNA                             |6    |\n",
      "|snoRNA                            |6    |\n",
      "|transcribed_processed_pseudogene  |4    |\n",
      "|unitary_pseudogene                |2    |\n",
      "|IG_J_gene                         |4    |\n",
      "|transcribed_unprocessed_pseudogene|40   |\n",
      "|pseudogene                        |1    |\n",
      "|IG_C_gene                         |4    |\n",
      "|sense_intronic                    |10   |\n",
      "|processed_pseudogene              |74   |\n",
      "|unprocessed_pseudogene            |72   |\n",
      "|miRNA                             |12   |\n",
      "|lincRNA                           |102  |\n",
      "+----------------------------------+-----+\n",
      "only showing top 20 rows\n",
      "\n",
      "+----------------------------------+-----+\n",
      "|consequenceTypesCounts            |value|\n",
      "+----------------------------------+-----+\n",
      "|missense_variant                  |4    |\n",
      "|intergenic_variant                |222  |\n",
      "|2KB_downstream_variant            |146  |\n",
      "|3_prime_UTR_variant               |9    |\n",
      "|regulatory_region_variant         |764  |\n",
      "|downstream_gene_variant           |163  |\n",
      "|intron_variant                    |543  |\n",
      "|splice_region_variant             |2    |\n",
      "|upstream_gene_variant             |200  |\n",
      "|TF_binding_site_variant           |108  |\n",
      "|5_prime_UTR_variant               |1    |\n",
      "|non_coding_transcript_exon_variant|45   |\n",
      "|non_coding_transcript_variant     |385  |\n",
      "|2KB_upstream_variant              |146  |\n",
      "|splice_donor_variant              |1    |\n",
      "|NMD_transcript_variant            |140  |\n",
      "|synonymous_variant                |6    |\n",
      "+----------------------------------+-----+\n",
      "\n",
      "+---------------+-----------+\n",
      "|chromosomeStats|value      |\n",
      "+---------------+-----------+\n",
      "|22             |[1000, 0.0]|\n",
      "+---------------+-----------+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "globalStatsTransformationDataframe = oskar.globalStats(df)\n",
    "globalStatsTransformationDataframe.select(explode(\"variantTypeCounts\").alias(\"variantTypeCounts\", \"value\")).show()\n",
    "globalStatsTransformationDataframe.select(explode(\"variantBiotypeCounts\").alias(\"variantBiotypeCounts\", \"value\")).show(truncate = False)\n",
    "globalStatsTransformationDataframe.select(explode(\"consequenceTypesCounts\").alias(\"consequenceTypesCounts\", \"value\")).show(truncate = False)\n",
    "globalStatsTransformationDataframe.select(explode(\"chromosomeStats\").alias(\"chromosomeStats\", \"value\")).show(truncate = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Histogram\n",
    "This next example shows how easily we can complement PyOskar API functionality with some well known python libraries like _pandas_ or _plotly_. We would like to know the frequency with which we can find variants depending on the region of our genome. For that task we will use _histogram_ PyOskar transformation, and after it we will store our transformed dataframe into a Pandas DataFrame with the default _toPandas()_ PySpark method so we can plot it later.\n",
    "<br>\n",
    "Usage:\n",
    "```python\n",
    "histogram(df[DataFrame], inputCol[str], step[int]=None)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------+-----+\n",
      "|   start|count|\n",
      "+--------+-----+\n",
      "|16000000|  101|\n",
      "|17000000|  100|\n",
      "|18000000|   85|\n",
      "|19000000|   53|\n",
      "|20000000|   46|\n",
      "|21000000|   85|\n",
      "|22000000|   52|\n",
      "|23000000|   49|\n",
      "|24000000|   44|\n",
      "|25000000|   62|\n",
      "|26000000|   42|\n",
      "|27000000|   54|\n",
      "|28000000|   21|\n",
      "|29000000|   29|\n",
      "|30000000|   38|\n",
      "|31000000|   46|\n",
      "|32000000|   44|\n",
      "|33000000|   34|\n",
      "|34000000|   15|\n",
      "+--------+-----+\n",
      "\n"
     ]
    }
   ],
   "source": [
    "oskar.histogram(df,\"start\",1000000).show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script><script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window._Plotly) {require(['plotly'],function(plotly) {window._Plotly=plotly;});}</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<script type=\"text/javascript\">window.PlotlyConfig = {MathJaxConfig: 'local'};</script><script type=\"text/javascript\">if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}</script><script>requirejs.config({paths: { 'plotly': ['https://cdn.plot.ly/plotly-latest.min']},});if(!window._Plotly) {require(['plotly'],function(plotly) {window._Plotly=plotly;});}</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "linkText": "Export to plot.ly",
        "plotlyServerURL": "https://plot.ly",
        "showLink": true
       },
       "data": [
        {
         "type": "bar",
         "uid": "f89b8c2d-443e-42ca-90b7-b2555e911f9d",
         "x": [
          16000000,
          17000000,
          18000000,
          19000000,
          20000000,
          21000000,
          22000000,
          23000000,
          24000000,
          25000000,
          26000000,
          27000000,
          28000000,
          29000000,
          30000000,
          31000000,
          32000000,
          33000000,
          34000000
         ],
         "y": [
          101,
          100,
          85,
          53,
          46,
          85,
          52,
          49,
          44,
          62,
          42,
          54,
          21,
          29,
          38,
          46,
          44,
          34,
          15
         ]
        }
       ],
       "layout": {
        "autosize": true,
        "xaxis": {
         "autorange": true,
         "range": [
          15500000,
          34500000
         ],
         "type": "linear"
        },
        "yaxis": {
         "autorange": true,
         "range": [
          0,
          106.3157894736842
         ],
         "type": "linear"
        }
       }
      },
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABQkAAAFoCAYAAAD0GX0ZAAAgAElEQVR4nO3d33Pd9X3n8f1PerU3nV7tpDd7se3OnPH6KHZaj7QsAc+E2Okwtae7GjODmVWcyFXiMWZhi7ODEaGlwdk6iQm0NALGMWQhJpFLAQMBTOQ4iSmk/IgsYyODbOu9F6lPZKzvVz6Wvzl6n8/jMfO6kI585vv11zJfPa0j/l0AAAAAAEX7d70+AAAAAACgt0RCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOGKiITnzp2Lu+66K3bv3n3ZYzMzM7Ft27YYGBiIoaGh2LdvX1ePAwAAAEB2fR8Jn3322bjhhhvi05/+9KKRcGxsLEZHR2N6ejqOHj0ag4ODMTk5ecWPAwAAAEB2fR8JL7r33nsvi4Rzc3PRbrfj2LFjnfeNj4/Hzp07r+hxAAAAAOgHRUfCEydORKvVirNnz3beNzExEZs3b76ix6ucOnsufm/r99Lua09OXcPfeQAAAABWuqIj4dTUVLRarZifn++87+DBg7Fhw4Yrejwi4tSpU5ftzXd+3fPQt5z9r8d+suh5Ve3Fn/0qDr5yIuWeefXNrs7VrBc7ffp0z4/BzPp7Z86c6fkxmFl/z98zZva7GMtTdCS8+J2Cc3NznfdNTEzEpk2brujxiIjz589ftukzH/U89C1nd3//jUXPq2r/fd8LPT/mq90ffuX7XZ2rWS924cKFnh+DmfX3/D1jZk3P3zNm9rsYy1N0JJybm4vVq1df8jMH9+zZEzt27Liix6uU9nLj//GtF3t+zMuJhAAAAAClKzoSRkSMjo7G9u3b4+TJkzE1NRVDQ0Nx6NChK358MSJhnomEAAAAAAVEwqeeeirWrVsXAwMD0W63Y926dfH00093Hp+eno6RkZFot9sxODgYe/fuveTXL/X4YkTCPBMJAQAAAAqIhL0gEuaZSAgAAAAgEjZCJMwzkRAAAABAJGyESJhnIiEAAACASNgIkTDPREIAAAAAkbARImGeiYQAAAAAImEjRMI8EwkBAAAARMJGiIR5JhICAAAAiISNEAnzTCQEAAAAEAkbIRLmmUgIAAAAIBI2QiTMM5EQAAAAQCRshEiYZyIhAAAAgEjYCJEwz0RCAAAAAJGwESJhnomEAAAAACJhI0TCPBMJAQAAAETCRoiEeSYSAgAAAIiEjRAJ80wkBAAAABAJGyES5plICAAAACASNkIkzDOREAAAAEAkbIRImGciIQAAAIBI2AiRMM9EQgAAAACRsBEiYZ6JhAAAAAAiYSNEwjzrNhI+/8uTPT/m5ezxV37V1fnSv35/2+M9//N4tdv29z/p9W8fAABA3xEJGyAS5plISKlEQgAAABYSCRsgEuaZSEipREIAAAAWEgkbIBLmmUhIqURCAAAAFio+Er755puxdevWuO666+L666+P3bt3x9zcXOfxmZmZ2LZtWwwMDMTQ0FDs27dvyecUCfNMJKRUIiEAAAALFR8JN27cGPv374+IiDNnzsTw8HA8+OCDncfHxsZidHQ0pqen4+jRozE4OBiTk5O1zykS5plISKlEQgAAABYqOhJeuHAhVq1aFUeOHOm87957741du3ZFRMTc3Fy02+04duxY5/Hx8fHYuXNn7fOKhHkmElIqkRAAAICFio6EEb+JfoODg/Hoo4/GsWPHYsOGDZ0oeOLEiWi1WnH27NnOx09MTMTmzZtrn1MkzDORkFKJhAAAACxUfCR8+eWX4+abb46dO3fGmjVrYnR0NGZnZyMiYmpqKlqtVszPz3c+/uDBg7Fhw4bO29PT05ftl796r+dfRC9nu7738qLnVbU/f/Bwz4/5avepsQNdnesPXvllz495Odv/46muztdWxmZmZq75c2aOhLd++/meXxOzftvp06d7fgxm1t87depUz4/BzPp/LE/RkfDUqVPxmc98Jt56662IiPjggw/i1ltvjdHR0Yj47XcSLvwfmUxMTMSmTZs6b8/Pz1+2mdm5nn8RvZztPvjTRc+ratm/k7Cbc/3nX0z3/JiXs8defrur87WVs2stcyT0nYQAkE+v76XMrIyxPEVHwldeeSXWrFlzyfsOHjwY69evj4jf/EzC1atXX/IzCffs2RM7duyofV4vN84zLzemVCIhAAAACxUdCc+cORPr1q2L/fv3x/z8fMzOzsbIyEjcfffdnY8ZHR2N7du3x8mTJ2NqaiqGhobi0KFDtc8rEuaZSEipREIAAAAWKjoSRkS8/vrrMTw8HENDQ3HDDTfEPffcEx999FHn8enp6RgZGYl2ux2Dg4Oxd+/eJZ9TJMwzkZBSiYQAAAAsVHwkbIJImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJGyASJhnIiGlEgkBAABYSCRsgEiYZyIhpRIJAQAAWEgkbIBImGciIaUSCQEAAFhIJIyI1157LbZu3Rpr166Ndrsd7733XuexmZmZ2LZtWwwMDMTQ0FDs27dvyecTCfNMJKRUIiEAAAALFR8JX3311RgcHIyJiYl4991345133om5ubnO42NjYzE6OhrT09Nx9OjRGBwcjMnJydrnFAnzTCSkVCIh5PLMT9/r+efecvbMT99b+iQBAOip4iPh8PBwPPbYY4s+Njc3F+12O44dO9Z53/j4eOzcubP2OUXCPBMJKZVICLmIhAAANK3oSHjmzJlotVoxNjYWQ0NDsWbNmti+fXucOXMmIiJOnDgRrVYrzp492/k1ExMTsXnz5s7b8/Pzl21mdq7nN+PL2e6DP130vKqWPRJ2c67//Ivpnh/zcvbYy293db7Wv8seCXv9+2f2u97Tb7zb88+95ezpN97t+e+hmZmZ9f9YnqIj4fHjx6PVasXk5GTMzs7Ge++9F7fcckuMjY1FRMTU1FS0Wq1L/qAdPHgwNmzY0Hn75MmTl+2Xv8r9r/13TLyy6HlV7c8fPNzzY77afWrsQFfn+v9+cqLnx7ycPTR5rKvztZWxU6dOXfPnzBwJb/328z2/Jma/6z32ws97/rm3nD32ws97/ntoZr3dBx980PNjMLP+H8sjErZaceHChc77nnvuuVi7dm1E/PY7CRf+jMKJiYnYtGlT7fN6uXGeebkxpcocCb3cmBJ5uTEAAE0rOhLOzs7GqlWr4vjx4533TU5Oxo033hgRv/mZhKtXr77kZxLu2bMnduzYUfu8ImGeiYSUSiSEXERCAACaVnQkjIj46le/Glu2bInp6el4//33Y3h4OB544IHO46Ojo7F9+/Y4efJkTE1NxdDQUBw6dKj2OUXCPBMJKZVICLmIhAAANK34SPjhhx/GHXfcEevWrYvrrrsu7r///jh37lzn8enp6RgZGYl2ux2Dg4Oxd+/eJZ9TJMwzkZBSiYSQi0gIAEDTio+ETRAJ80wkpFQiIeQiEgIA0DSRsAEiYZ6JhJRKJIRcREIAAJomEjZAJMwzkZBSiYSQi0gIAEDTRMIGiIR5JhJSKpEQchEJAQBomkjYAJEwz0RCSiUSQi4iIQAATRMJGyAS5plISKlEQshFJOxfL5w4GXcdeCPtPjh7rte/hQDANSISNkAkzDORkFKJhJCLSNi//ubQz3t+fZazt2fO9vq3EAC4RkTCBoiEeSYS1vufD78c/238xyk39r1XuzrX0oiEkItI2L9EQgBgpRAJGyAS5plIWO8zXzvU82O+2n32vh93da6lEQkhF5Gwf4mEAMBKIRI2QCTMM5GwnkjYv0qKhDsff73nx7ycnb8w39CfAjIRCfuXSAgArBQiYQNEwjwTCeuJhP1LJMwzkZAIkbCfiYQAwEohEjZAJMwzkbCeSNi/RMI8EwmJEAn7mUgIAKwUImEDRMI8EwnriYT9SyTMM5GQCJGwn4mEAMBKIRI2QCTMM5GwnkjYv0TCPBMJiRAJ+5lICACsFCJhA0TCPBMJ64mE/UskzLNuIuHbM2d7frzL2d8c+nmXf5LLIRL2L5EQAFgpRMIGiIR5JhLWEwn7l0iYZyIhESJhPxMJAYCVQiRsgEiYZyJhPZGwf4mEeSYSEiES9jOREABYKUTCBoiEeSYS1hMJ+5dImGciIREiYT8TCQGAlUIkbIBImGciYT2RsH+JhHkmEhIhEvYzkRAAWClEwgaIhHkmEtYrKRLu/+c3e37My9nUO6e7Ol+RMM9EQiJEwn4mEkI+n73vxz3/3LvafeZrh3r92wesYCJhA0TCPBMJ64mEeSYSVhMJ80wkrCYS9i+REPIRCYF+JRI2QCTMM5GwnkiYZyJhNZEwz0TCaiJh/xIJIR+REOhXImEDRMI8EwnriYR5JhJWEwnzrNtI+A9H3oq7DryRcg882925ioT9SySEfEqKhFPvnO75fzOXM39HQXdEwgaIhHkmEtYTCfNMJKwmEuZZt5HwC994rufHfLX74zt+0NW5ioT9SySEfEqKhI+/8queH/Ny9vwvTzb0pwD6k0j4bz7++OPYsmVLbNmy5ZL3z8zMxLZt22JgYCCGhoZi3759Sz6XSJhnImE9kTDPRMJqImGeiYTVRML+JRJCPiJhnomE0B2RMCIuXLgQX/7yl2Pz5s2XRcKxsbEYHR2N6enpOHr0aAwODsbk5GTt84mEeSYS1hMJ80wkrCYS5plIWE0k7F8iIeQjEuaZSAjdEQkj4s4774zx8fF44oknLomEc3Nz0W6349ixY533jY+Px86dO2ufTyTMM5GwnkiYZyJhNZEwz0TCaqVFwj++4wc9P+ar3Re+8VxX5yoSQj4iYZ6JhNCd4iPh/fffH7fffnvMz89fFglPnDgRrVYrzp797c3PxMREbN68ufY5RcI8EwnriYR5JhJWEwnzTCSsJhLmmUgI/U8kzDORELpTdCQ8cOBAjIyMxPnz5yMiLouEU1NT0Wq1Yn7+t1+gHTx4MDZs2NB5+9y5c5ft16dzf5F29/ffWPS8qvbf973Q82O+2v3hV74fc3NzV7zDP8v9Rdo/vvhmV+e7dvcPe37MV7vrx3/U1bl+6/Aven7My9nrb53s6vM2cyQcefjlrs51x8SrPT/m5eyjj+eu+FxPvH+658e7nN3/zM+6urYb//afen7MV7s/vuMHXZ3rD17/154f83L2g9f/tavzzRwJN/7tP3V1rvc/87OeH/NyduL9012dr5W9bu7NVvKuH/9Rzz/3rnZrd/+wq3P9xxdz/0P64Z+91/M/L/ab/a7+nmF5io6E9913X6xevTra7Xa02+1YtWpVtFqtaLfbcfbs2c53Es7NzXV+zcTERGzatKnz9qlTpy7bm+/8uud/GS5n/+uxnyx6XlXbtDfvF2mfGjsQH3zwwRXvh6+/1fNjXs4e/qfjXZ3vp//q6Z4f89Xuuj2HujrXvYemen7My9mR4//a1flmjoS37X+xq3P9y394qefHvJydnDl1xef603/J/Q8Ze5482tW1vemv834nx3+6/cmuzvWJI7/s+TEvZ08c+WVX5/ufbn+y58d8tbvpr3/c1bnuefJoz495Ofvpv7zX1flauTt9+nTPj+Fa7bo9eV9t8+m/erqrc334n473/JiXsx++/lbP/7zYb9ZNY1jOWJ6iI+EnLfYzCVevXn3JzyTcs2dP7Nixo/Z5vNw4z7zcuJ6XG+eZlxtX83LjPPNy42pebpxnXm4M/c/LjfPMy42hOyLhAp+MhBERo6OjsX379jh58mRMTU3F0NBQHDpU/xerSJhnImE9kTDPRMJqImGeiYTVRMI8Ewmh/4mEeSYSQndEwgUWi4TT09MxMjIS7XY7BgcHY+/evUs+j0iYZyJhPZEwz0TCaiJhnomE1UTCPBMJKdHs3Pn40c/eT7t3P/ioq/MVCfNMJITuiIQNEAnzTCSsJxLmmUhYTSTMM5GwmkiYZyIhJZp6J/f/OGv/P7/Z1fmKhHkmEkJ3RMIGiIR5JhLWEwnzTCSsJhLmmUhYTSTMM5GQEomEeSYSAnVEwgaIhHkmEtYTCfNMJKwmEuaZSFhNJMwzkZASiYR5JhICdUTCBoiEeSYS1hMJ80wkrCYS5plIWE0kzDORkBKJhHkmEgJ1RMIGiIR5JhLWEwnzTCSsJhLmmUhYTSTMM5GQEomEeSYSAnVEwgaIhHkmEtYTCfNMJKwmEuaZSFhNJMwzkZASiYR5JhICdUTCBoiEeSYS1hMJ80wkrCYS5plIWE0kzDORkBKJhHkmEta77+mfxV0H3ki5br/Wg8WIhA0QCfNMJKwnEuaZSFhNJMwzkbCaSJhnIiElEgnzTCSs94df+X7Pj/lq9z++9WJX5wqLEQkbIBLmmUhYTyTMM5GwmkiYZyJhNZEwz0RCSiQS5plIWE8kpHQiYQNEwjwTCeuJhHkmElYTCfNMJKwmEuaZSEiJRMI8EwnriYSUTiRsgEiYZyJhPZEwz0TCaiJhnomE1UTCPBMJKZFImGciYT2RkNKJhA0QCfNMJKwnEuaZSFhNJMwzkbCaSJhnIiElEgnzTCSsJxJSOpGwASJhnomE9UTCPBMJq4mEeSYSVhMJ80wkpEQiYZ6JhPVEQkonEjZAJMwzkbCeSJhnImE1kTDPRMJqImGeiYSUSCTMM5GwnkhI6UTCBoiEeSYS1hMJ80wkrCYS5plIWE0kzDORkBKJhHkmEtYTCSmdSNgAkTDPRMJ6ImGeiYTVRMI8EwmriYR5JhJy0ba//0nPr8/V7ve3Pd7VuYqEeSYS1hMJKZ1I2ACRMM9EwnoiYZ6JhNVEwjwTCauJhHkmEnKRSJhnImE1kTDPREKuBZGwASJhnomE9UTCPBMJq4mEeSYSVhMJ80wk5CKRMM9EwmoiYZ6JhFwLImEDRMI8EwnriYR5JhJWEwnzTCSsJhLmmUjIRSJhnomE1UTCPBMJuRZEwgaIhHkmEtYTCfNMJKwmEuaZSFhNJMwzkZCLRMI8EwmriYR5JhJyLYiEDRAJ80wkrCcS5plIWE0kzDORsJpImGciIReJhHkmElYTCfNMJORaEAkbIBLmmUhYTyTMM5GwmkiYZyJhNZEwz0RCLhIJ80wkrCYS5plIyLUgEjZAJMwzkbCeSJhnImE1kTDPRMJqImGeiYRcJBLmmUhYTSTMM5GQa0EkbIBImGciYT2RMM9EwmoiYZ6JhNVEwjwTCblIJMwzkbCaSJhnIiHXgkjYAJEwz0TCeiJhnomE1UTCPBMJq4mEeSYScpFImGciYTWRMM9EQq6F4iPhAw88EJ/73Oei3W7H+vXr4+GHH77k8ZmZmdi2bVsMDAzE0NBQ7Nu3b8nnFAnzTCSsJxLmmUhYTSTMM5GwmkiYZyIhF4mEeSYSVhMJ80wk5FooPhKOj4/HkSNH4uTJk3H48OEYGBiIF154ofP42NhYjI6OxvT0dBw9ejQGBwdjcnKy9jlFwjwTCeuJhHkmElYTCfNMJKwmEuaZSMhFImGeiYTVRMI8Ewm5FoqPhJ80PDwcDz30UEREzM3NRbvdjmPHjnUeHx8fj507d9Y+h0iYZyJhPZEwz0TCaiJhnomE1UTCPBMJuUgkzDORsJpImGciIdeCSLjA2bNnY926dfH8889HRMSJEyei1WrF2bO/vfmZmJiIzZs31z6PSJhnImE9kTDPRMJqImGeiYTVRMI8Ewnr3XXgjbT74VR3f45FwjwTCauJhHkmEnItiIQL7Nq1K7Zu3Rrz87/5gmxqaiparVbn7YiIgwcPxoYNGzpvf/zxx5ftvVMf9vwviOXsrw4cXfS8qvYXf/d8z4/5aveHX/l+V+c6eezdnh/zcvaPL77Z1fmu3f3Dnh/z1e768R91da77JnN/kfbav0x3db6ZI+HIwy93da5f/d5Pen7My9ns2Y+u+Fx/8e6pnh/vcvb1p491dW03PHC458d8tfujXU91da5Pvvp2z495OXvy1be7Ot8/2vVUz4/5arfhgcNdnevXnz7W82Nezn7x7qkrPtfZsx/1/HiXs69+7yddXduRh1/u+TFf7X5/2+Ndnetr/zLd82NezvZN/ryr871+/Ec9P+ar3drdP+zqXP/xxdz/kD557N2uzjdzJPyLv3u+q3Pt17E8IuG/ueeee+Lmm2+OM2fOdN538TsJ5+bmOu+bmJiITZs2dd4+c+bMZXv7/Zme/wWxnN35+KuLnlfVNn8z73dyfGrsQFfneuho7i/SHnnu512d75q7n+n5MV/trttzqKtz/eazub9Ie+nn73R1vpkj4W37X+zqXMcezftF2u9t/V6c+uD0FZ/rsbd/3fPjXc7ufeqNrq7t5/9msufHfLX7o11PdXWuB17O/UXagZff7Op8M0fCz//NZFfneu9Tb/T8mJezY2//+orP9dQHub/bbOzRl7u6trftz/tqm9/f9nhX5/rSz9/p+TEvZ9989lhX53vdnryvtllz9zNdnesjz+X+h/RDR9/u6nw/NXag58d8tdv8zee6Otd+HctTfCQ8f/587Nq1K4aHhy/7AzU3NxerV6++5GcS7tmzJ3bs2FH7nF5unGdeblzPy43zzMuNq3m5cZ55uXE1LzfOMy83rnb+wnzPj3c52/n4611dWy83zjMvN67m5cZ55uXGXAtFR8Lz58/HrbfeGl/84hdjZmYmZmdnY3Z29pKfQTg6Ohrbt2+PkydPxtTUVAwNDcWhQ/V/sYqEeSYS1hMJ80wkrCYS5plIWE0kzDORsJpImGciYT2RMM9EQuhO0ZHw9OnT0Wq1LttNN93U+Zjp6ekYGRmJdrsdg4ODsXfv3iWfVyTMM5GwnkiYZyJhNZEwz0TCaiJhnomE1UTCPBMJ64mEeSYSQneKjoRNEQnzTCSsJxLmmUhYTSTMM5GwmkiYZyJhNZEwz0TCeiJhnomE0B2RsAEiYZ6JhPVEwjwTCauJhHkmElYTCfNMJKwmEuaZSFhPJMwzkRC6IxI2QCTMM5GwnkiYZyJhNZEwz0TCaiJhnomE1UTCPBMJ64mEeSYSQndEwgaIhHkmEtYTCfNMJKwmEuaZSFhNJMwzkbCaSJhnImE9kTDPRELojkjYAJEwz0TCeiJhnomE1UTCPBMJq4mEeSYSVhMJ80wkrCcS5plIWO3w8V/HXQfeSDuaIRI2QCTMM5GwnkiYZyJhNZEwz0TCaiJhnomE1UTCPBMJ64mEeSYSVvvak1M9P+bl7NTZc12dL1dGJGyASJhnImE9kTDPRMJqImGeiYTVRMI8EwmriYR5JhLWEwnzTCSsJhKyGJGwASJhnomE9UTCPBMJq4mEeSYSVhMJ80wkrCYS5plIWE8kzDORsJpIyGJEwgaIhHkmEtYTCfNMJKwmEuaZSFhNJMwzkbCaSJhnImE9kTDPRMJqIiGLEQkbIBLmmUhYTyTMM5GwmkiYZyJhNZEwz0TCaiJhnomE9UTCPBMJq4mELEYkbIBImGciYT2RMM9EwmoiYZ6JhNVEwjwTCauJhHkmEtYTCfNMJKwmErIYkbABImGeiYT1RMI8EwmriYR5JhJWEwnzTCSsJhLmmUhYTyTMM5GwmkjIYkTCBoiEeSYS1hMJ80wkrCYS5plIWE0kzDORsJpImGciYT2RMM9EwpxC9+IAAA34SURBVGoiIYsRCRsgEuaZSFhPJMwzkbCaSJhnImE1kTDPRMJqImGeiYT1RMI8EwmriYQsRiRsgEiYZyJhPZEwz0TCaiJhnomE1UTCPBMJq4mEeSYS1hMJ80wkrCYSshiRsAEiYZ6JhPVEwjwTCauJhHkmElYTCfNMJKwmEuaZSFhPJMwzkbCaSMhiRMIGiIR5JhLWEwnzTCSsJhLmmUhYTSTMM5GwmkiYZyJhPZEwz0TCaiIhixEJGyAS5plIWE8kzDORsJpImGciYTWRMM9EwmoiYZ6JhPVEwjwTCauJhCxGJGyASJhnImE9kTDPRMJqImGeiYTVRMI8EwmriYR5JhLWEwnzTCSsJhKyGJGwASJhnomE9UTCPBMJq4mEeSYSVhMJ80wkrCYS5plIWE8kzDORsJpIyGJEwgaIhHkmEtYTCfNMJKwmEuaZSFhNJMwzkbCaSJhnImE9kTDPRMJqIiGLEQkbIBLmmUhYTyTMM5GwmkiYZyJhNZEwz0TCaiJhnomE9UTCPBMJq4mELEYkbIBImGciYT2RMM9EwmoiYZ6JhNVEwjwTCauJhHkmEtYTCfNMJKwmErIYkbABImGeiYT1RMI8EwmriYR5JhJWEwnzTCSsJhLmmUhYTyTMM5GwmkjIYkTCBoiEeSYS1hMJ80wkrCYS5plIWE0kzDORsJpImGciYT2RMM9EwmoiIYsRCZcwMzMT27Zti4GBgRgaGop9+/Yt+WtEwjwTCeuJhHkmElYTCfNMJKwmEuaZSFhNJMwzkbCeSJhnImE1kZDFiIRLGBsbi9HR0Zieno6jR4/G4OBgTE5O1v4akTDPRMJ6ImGeiYTVRMI8EwmriYR5JhJWEwnzTCSsJxLmmUhYTSRkMSJhjbm5uWi323Hs2LHO+8bHx2Pnzp21v04kzDORsJ5ImGciYTWRMM9EwmoiYZ6JhNVEwjwTCeuJhHkmElYTCVmMSFjjxIkT0Wq14uzZ3978TExMxObNm2t/nUiYZyJhPZEwz0TCaiJhnomE1UTCPBMJq4mEeSYS1hMJ80wkrCYSshiRsMbU1FS0Wq2Yn//tF2gHDx6MDRs2dN5+8MEHL9v4A3t7/gmznG28c/+i51W1tTu+2/Njvtr9wcijXZ3rV+79u54f83J22//5Vlfn+x+//Pc9P+ar3X/+y0e6Otfhu7/V82Nezu78+v/t6nz//W2P9vyYr3b/dedDXZ3r+jv29/yYl7O//caVn+vX/vqbPT/e5ezP//d3urq2/+UrD/f8mK92/2HbP3R1rl+6Z1/Pj3k5+9I9+7o63/+w7R96fsxXu//ylYe7Otc//9/f6fkxL2df++tvXvG5/u03Huz58S5n6+/o7h75v+58qOfHfLX797d1d49859f/b8+PeTkbvru7e+T//JeP9PyYr3b/8ct/39W53vZ/ct8jf+Xev+vqfP9gJO898tod3+3qXDfemfseefyBvYueF8sjEta4+J2Ec3NznfdNTEzEpk2bOm9/4xvfMDNrdN/+9rd7fgxm1t/77ne/2/NjMLP+nvsZM/tdjOURCWvMzc3F6tWrL/mZhHv27IkdO3b08KgAAAAA4NoSCZcwOjoa27dvj5MnT8bU1FQMDQ3FoUPd/RwHAAAAAFjJRMIlTE9Px8jISLTb7RgcHIy9e/f2+pAAAAAA4JoSCVeQc+fOxV133RW7d+9e9PHXXnsttm7dGmvXro12ux3vvffb/1Pgxo0b40//9E9jdna2877z58/HDTfcEH/yJ3/S+LFTr+7avvnmm7F169a47rrr4vrrr4/du3df8nMwXduV6YEHHojPfe5z0W63Y/369fHwww9f8vjMzExs27YtBgYGYmhoKPbt23fJ467ryrXUtV3qcdd25Vrq2l308ccfx5YtW2LLli2XvN+1Xbmu5Nq6j8ppqWvrPiqv73znO/GFL3whBgYGYt26dXH77bfHhx9+2HncvVROS11X91F5LXVtL3IflZtIuEI8++yzccMNN8SnP/3pRUPSq6++GoODgzExMRHvvvtuvPPOO5fdAN14442xf//+zvuefPLJWL9+vU+4Hlvq2m7cuLFz3c6cORPDw8OX/F+ZXNuVaXx8PI4cORInT56Mw4cPx8DAQLzwwgudx8fGxmJ0dDSmp6fj6NGjMTg4GJOTk53HXdeVa6lru9Tjru3KtdS1i4i4cOFCfPnLX47NmzcvenPr2q5MS11b91F5Xcnfue6jcnriiSfi9ddfj1OnTsUvfvGL+PznPx/f/va3O4+7l8ppqevqPiqvpa5thPuofiASrjD33nvvoiFpeHg4Hnvsscpft3HjxnjkkUfis5/9bJw7dy4iIjZt2hSPPPKIT7gVYrFre+HChVi1alUcOXLkko/btWtX523XNofh4eF46KGHIuI3/9Ojdrt9yf/0aHx8PHbu3Nl523XNY+G1vZLHXds8Fru2d955Z4yPj8cTTzyx6M2ta5vDJ6+t+6j+sfDauo/qDxcuXIjjx4/HTTfdFC+++GJEuJfqB4td18W4j8qn7tq6j8pPJFxhFgtJZ86ciVarFWNjYzE0NBRr1qyJ7du3x5kzZzofs3Hjxjhy5Ejcdttt8cQTT8RLL70Ut9xyS7z22ms+4VaIqgA8Pj4eg4OD8eijj8axY8diw4YNl9wQubYr39mzZ2PdunXx/PPPR0TEiRMnotVqxdmzZzsfMzExEZs3b+687brm8MlreyWPu7Y5LHbt7r///rj99ttjfn6+8ubWtV35Pnlt3Uf1j8U+b91H5ddqtWL16tWXhHz3Uvktdl0/yX1UTlXX1n1UfxAJV5jFQtLx48ej1WrF5ORkzM7OxnvvvRe33HJLjI2NdT7m4ifcCy+8EBs2bIiRkZE4fPiwT7gVpCoSvvzyy3HzzTfHzp07Y82aNTE6OnrJz2lwbVe+Xbt2xdatW2N+fj4iIqampqLVanXejog4ePBgbNiwofO265rDJ6/tlTzu2ubwyWt34MCBGBkZifPnz0dE1N7curYr2yevrfuo/rHY37nuo/K7+F1JN954YzzzzDMR4V6qHyx2XT/JfVROi11b91H9QyRcYeoi4YULFzrve+6552Lt2rWdty9+wkX85lt2/+zP/iwiwifcCrLYtT116lR85jOfibfeeisiIj744IO49dZbY3R0tPMxru3Kds8998TNN998yXekXPzX74U/72piYiI2bdrUedt1XfkWu7ZX8rhru/Itdu3uu+++WL16dbTb7Wi327Fq1apotVrRbrc738ni2q58i11b91H9YbFr6z6qv3z961/vxHv3Uv1j4XVdyH1Ufguvrfuo/iESrjCLhaTZ2dlYtWpVHD9+vPO+ycnJuPHGGztvL/yEe+WVV+Kll16KCJ9wK8li1/aVV16JNWvWXPK+gwcPxvr16ztvu7Yr0/nz52PXrl0xPDx82c3N3NxcrF69+pKXO+3Zsyd27NjRedt1Xbnqru2VPO7arlxLXbuF6v4FPMK1XWnqrq37qNzqrq37qP5y9913x+233x4R7qX6ycLrGuE+qp988tou5D4qL5Fwhal6SepXv/rV2LJlS0xPT8f7778fw8PD8cADD3QeX/gJt5BPuJWj6udNrlu3Lvbv3x/z8/MxOzsbIyMjcffdd3c+xrVdec6fPx+33nprfPGLX4yZmZmYnZ2N2dnZS35uzujoaGzfvj1OnjwZU1NTMTQ0FIcOHeo87rquTEtd2yu59q7tynQl126hpW5uF3Jte+tKrq37qJyWurbuo/K6cOFCfOlLX4oXX3wxpqen45lnnok1a9bE4cOHOx/jXiqfpa6r+6i8ruRzdiH3UXmJhCvEU089FevWrYuBgYFot9uxbt26ePrppzuPf/jhh3HHHXfEunXr4rrrrov777+/838EivAJt5ItdW1ff/31GB4ejqGhobjhhhvinnvuiY8++qjzuGu78pw+fTpardZlu+mmmzofMz09HSMjI9Fut2NwcDD27t17yXO4rivTUtf2Sq69a7syXcm1W8jNbR5Xcm3dR+V0JdfWfVRO8/Pzcccdd8T69euj3W7Hhg0b4sCBA5d8jHupfJa6ru6j8rqSz9mF3EflJRICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhRMJAQAAAKBwIiEAAAAAFE4kBAAAAIDCiYQAAAAAUDiREAAAAAAKJxICAAAAQOFEQgAAAAAonEgIAAAAAIUTCQEAAACgcCIhAAAAABROJAQAAACAwomEAAAAAFA4kRAAAAAACicSAgAAAEDhREIAAAAAKJxICAAAAACFEwkBAAAAoHAiIQAAAAAUTiQEAAAAgMKJhAAAAABQOJEQAAAAAAonEgIAAABA4URCAAAAACicSAgAAAAAhfv/yhwjfdCwcaIAAAAASUVORK5CYII=",
      "text/html": [
       "<div id=\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\", [{\"x\": [16000000, 17000000, 18000000, 19000000, 20000000, 21000000, 22000000, 23000000, 24000000, 25000000, 26000000, 27000000, 28000000, 29000000, 30000000, 31000000, 32000000, 33000000, 34000000], \"y\": [101, 100, 85, 53, 46, 85, 52, 49, 44, 62, 42, 54, 21, 29, 38, 46, 44, 34, 15], \"type\": \"bar\", \"uid\": \"c5ccc0ec-e292-4337-8a95-74383d5d7d45\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\"})});</script><script type=\"text/javascript\">window.addEventListener(\"resize\", function(){window._Plotly.Plots.resize(document.getElementById(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\"));});</script>"
      ],
      "text/vnd.plotly.v1+html": [
       "<div id=\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\" style=\"height: 525px; width: 100%;\" class=\"plotly-graph-div\"></div><script type=\"text/javascript\">require([\"plotly\"], function(Plotly) { window.PLOTLYENV=window.PLOTLYENV || {};window.PLOTLYENV.BASE_URL=\"https://plot.ly\";Plotly.newPlot(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\", [{\"x\": [16000000, 17000000, 18000000, 19000000, 20000000, 21000000, 22000000, 23000000, 24000000, 25000000, 26000000, 27000000, 28000000, 29000000, 30000000, 31000000, 32000000, 33000000, 34000000], \"y\": [101, 100, 85, 53, 46, 85, 52, 49, 44, 62, 42, 54, 21, 29, 38, 46, 44, 34, 15], \"type\": \"bar\", \"uid\": \"c5ccc0ec-e292-4337-8a95-74383d5d7d45\"}], {}, {\"showLink\": true, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\"})});</script><script type=\"text/javascript\">window.addEventListener(\"resize\", function(){window._Plotly.Plots.resize(document.getElementById(\"c95cacd7-e4d8-4af1-adf6-05baf8ca85d9\"));});</script>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly.offline as py\n",
    "import plotly.figure_factory as ff\n",
    "import plotly.graph_objs as go\n",
    "from plotly.offline import get_plotlyjs, init_notebook_mode\n",
    "\n",
    "pandasHistogramDF = oskar.histogram(df,\"start\",1000000).toPandas()\n",
    "\n",
    "init_notebook_mode(connected = True)\n",
    "\n",
    "data = [go.Bar(x= pandasHistogramDF[\"start\"], y = pandasHistogramDF[\"count\"])]\n",
    "py.iplot(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "** If you are using jupyter lab you will need to download this extension https://github.com/jupyterlab/jupyter-renderers/tree/master/packages/plotly-extension to visualize the plot*"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
